% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/regression-pipeline.R
\name{iq_regression}
\alias{iq_regression}
\title{Perform IQ regression on peak intervals}
\usage{
iq_regression(
  peak_intervals,
  atac_scores = NULL,
  atac_diff = NULL,
  normalize_bins = TRUE,
  norm_intervals = NULL,
  motif_energies = NULL,
  additional_features = NULL,
  add_sequences_features = TRUE,
  max_motif_num = 30,
  traj_prego = NULL,
  peaks_size = 500,
  bin_start = 1,
  bin_end = NULL,
  seed = 60427,
  frac_train = 0.8,
  train_idxs = NULL,
  test_idxs = NULL,
  filter_model = TRUE,
  min_diff = 0.1,
  prego_min_diff = min_diff,
  prego_sample_for_kmers = TRUE,
  prego_sample_fraction = 0.1,
  prego_energy_norm_quantile = 1,
  prego_spat_bin_size = NULL,
  prego_spat_num_bins = NULL,
  r2_threshold = 0.0005,
  bits_threshold = 1.75,
  filter_sample_frac = 0.1,
  include_interactions = FALSE,
  interaction_threshold = 0.001,
  max_motif_interaction_n = NULL,
  max_add_interaction_n = NULL,
  n_prego_motifs = 0,
  n_cores = NULL,
  output_dir = NULL,
  ...
)
}
\arguments{
\item{peak_intervals}{A data frame, indicating the genomic positions ('chrom', 'start', 'end') of each peak.}

\item{atac_scores}{Optional. A numeric matrix, representing mean ATAC score per bin per peak. Rows: peaks, columns: bins. By default iceqream would regress the last column minus the first column. If you want to regress something else, please either change bin_start or bin_end, or provide \code{atac_diff} instead. If \code{normalize_bins} is TRUE, the scores will be normalized to be between 0 and 1.}

\item{atac_diff}{Optional. A numeric vector representing the differential accessibility between the start and end of the trajectory. Either this or \code{atac_scores} must be provided.}

\item{normalize_bins}{whether to normalize the ATAC scores to be between 0 and 1. Default: TRUE}

\item{norm_intervals}{A data frame, indicating the genomic positions ('chrom', 'start', 'end') of peaks used for energy normalization. If NULL, the function will use \code{peak_intervals} for normalization.}

\item{motif_energies}{A numeric matrix, representing the energy of each motif in each peak. If NULL, the function will use \code{pssm_db} to calculate the motif energies. Note that this might take a while.}

\item{additional_features}{A data frame, representing additional genomic features (e.g. CpG content, distance to TSS, etc.) for each peak. Note that NA values would be replaced with 0.}

\item{add_sequences_features}{Add CG content and dinuceotide content to the additional features.}

\item{max_motif_num}{maximum number of motifs to consider. Default: 50}

\item{traj_prego}{output of \code{learn_traj_prego}. If provided, no additional prego models would be inferred.}

\item{peaks_size}{size of the peaks to extract sequences from. Default: 500bp}

\item{bin_start}{the start of the trajectory. Default: 1}

\item{bin_end}{the end of the trajectory. Default: the last bin (only used when atac_scores is provided)}

\item{seed}{random seed for reproducibility.}

\item{frac_train}{A numeric value indicating the fraction of intervals to use for training (default is 0.8).}

\item{train_idxs}{A vector of indices to use for training. If NULL, the training set is randomly selected.}

\item{test_idxs}{A vector of indices to use for testing. If NULL, the testing set is the complement of the training set.}

\item{filter_model}{A logical value indicating whether to filter the model (default is TRUE).}

\item{min_diff}{minimal ATAC difference for a peak to participate in the initial prego motif inference and in the distillation step (if \code{distill_on_diff} is TRUE).}

\item{prego_sample_for_kmers}{whether to use a sample of the peaks for kmer screening. Default: TRUE}

\item{prego_sample_fraction}{Fraction of peaks to sample for prego motif inference. A smaller number would be faster but might lead to over-fitting. Default: 0.1}

\item{r2_threshold}{minimal R^2 for a feature to be included in the model.}

\item{bits_threshold}{minimal sum of bits for a feature to be included in the model.}

\item{filter_sample_frac}{The fraction of samples to use for computing the r2 without each model at the filtering step. When NULL, all samples are used.}

\item{include_interactions}{whether to include interactions between motifs / additional fetures as model features. IQ will create interactions between significant additional features and all motifs, and between significant motifs. Default: FALSE}

\item{interaction_threshold}{threshold for the selecting features to create interactions. IQ learns a linear model on the features and selects the features with coefficients above this threshold. Default: 0.001}

\item{max_motif_interaction_n}{maximum number of motifs to consider for interactions. If NULL, all motifs above the interaction_threshold will be considered. Default: NULL}

\item{max_add_interaction_n}{maximum number of additional features to consider for interactions. If NULL, all additional features above the interaction_threshold will be considered. Default: NULL}

\item{n_prego_motifs}{number of prego motifs (de-novo motifs) to consider.}

\item{n_cores}{The number of cores to use for parallel processing. When NULL, the number of threads is automatically determined as 80\% of the available cores. See \code{\link{prego::set_parallel}} for more details.}

\item{output_dir}{A directory to save intermediate results. If not NULL, the train and test indices are saved to a CSV file, together with the models at each step (before filtering, after filtering, and after adding interactions). The models are saved as RDS files. If the directory exists, the files are overwritten.}

\item{...}{
  Arguments passed on to \code{\link[=regress_trajectory_motifs]{regress_trajectory_motifs}}
  \describe{
    \item{\code{n_clust_factor}}{factor to divide the number of to keep after clustering. e.g. if n_clust_factor > 1 the number of motifs to keep will be reduced by a factor of n_clust_factor. Default: 1}
    \item{\code{norm_motif_energies}}{A numeric matrix, representing the normalized energy of each motif in each interval of \code{norm_intervals}. If NULL, the function will use \code{pssm_db} to calculate the motif energies. Note that this might take a while.}
    \item{\code{pssm_db}}{a data frame with PSSMs ('A', 'C', 'G' and 'T' columns), with an additional column 'motif' containing the motif name. All the motifs in \code{motif_energies} (column names) should be present in the 'motif' column. Default: all motifs.}
    \item{\code{min_tss_distance}}{distance from Transcription Start Site (TSS) to classify a peak as an enhancer. Default: 5000. If NULL, no filtering will be performed - use this option if your peaks are already filtered. \cr
Note that in order to filter peaks that are too close to TSS, the current \code{misha} genome must have an intervals set called \code{intervs.global.tss}.}
    \item{\code{normalize_energies}}{whether to normalize the motif energies. Set this to FALSE if the motif energies are already normalized.}
    \item{\code{min_initial_energy_cor}}{minimal correlation between the motif normalized energy and the ATAC difference.}
    \item{\code{energy_norm_quantile}}{quantile of the energy used for normalization. Default: 1}
    \item{\code{norm_energy_max}}{maximum value of the normalized energy. Default: 10}
    \item{\code{distill_on_diff}}{whether to distill motifs based on differential accessibility. If FALSE, all peaks will be used for distillation, if TRUE - only peaks with differential accessibility >= min_diff will be used.}
    \item{\code{feature_selection_beta}}{beta parameter used for feature selection.}
    \item{\code{filter_using_r2}}{whether to filter features using R^2.}
    \item{\code{parallel}}{whether to use parallel processing on glmnet.}
    \item{\code{spat_num_bins}}{number of spatial bins to use.}
    \item{\code{spat_bin_size}}{size of each spatial bin.}
    \item{\code{kmer_sequence_length}}{length of the kmer sequence to use for kmer screening. By default the full sequence is used.}
    \item{\code{max_interaction_n}}{maximum number of interactions to consider.}
    \item{\code{alpha}}{The elasticnet mixing parameter, with \eqn{0\le\alpha\le 1}.
The penalty is defined as
\deqn{(1-\alpha)/2||\beta||_2^2+\alpha||\beta||_1.} \code{alpha=1} is the
lasso penalty, and \code{alpha=0} the ridge penalty.}
    \item{\code{lambda}}{A user supplied \code{lambda} sequence. Typical usage is to
have the program compute its own \code{lambda} sequence based on
\code{nlambda} and \code{lambda.min.ratio}. Supplying a value of
\code{lambda} overrides this. WARNING: use with care. Avoid supplying a
single value for \code{lambda} (for predictions after CV use
\code{predict()} instead).  Supply instead a decreasing sequence of
\code{lambda} values. \code{glmnet} relies on its warms starts for speed,
and its often faster to fit a whole path than compute a single fit.}
  }}
}
\value{
An instance of \code{TrajectoryModel} with the final model.
}
\description{
Perform IQ regression on peak intervals using the provided ATAC-seq scores, ATAC-seq score differences, normalized intervals, motif energies, and additional features, after dividing the intervals into training and testing sets.
}
