<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="iceqream">
<title>IceQream • iceqream</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="IceQream">
<meta property="og:description" content="iceqream">
<meta property="og:image" content="https://tanaylab.github.io/iceqream/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">iceqream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/iceqream.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">

<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>IceQream</h1>
            
      
      
      <div class="d-none name"><code>iceqream.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>IceQream (Interpretable Computational Engine for Quantitative Regression of Enhancer ATAC Motifs) is a package for regressing accessibility from DNA sequences. It models TF effective concentrations as latent variables that activate or repress regulatory elements in a nonlinear fashion, with possible contribution from pairwise interactions and synergistic chromosomal domain effects.</p>
<p>This vignette demonstrates how to use IceQream to analyze chromosome accessibility data, specifically focusing on a mouse gastrulation trajectory from Epiblast to Early nascent mesoderm. This analysis can help identify key regulatory elements and transcription factors involved in cellular differentiation processes, and understand quantitatively how they contribute to the observed changes in chromatin accessibility.</p>
</div>
<div class="section level2">
<h2 id="setup-and-data-preparation">Setup and Data Preparation<a class="anchor" aria-label="anchor" href="#setup-and-data-preparation"></a>
</h2>
<p>First, let's load the necessary packages and set up our environment:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tanaylab.github.io/iceqream/">iceqream</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tanaylab.github.io/misha/" class="external-link">misha</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">misha.ext</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'misha.ext'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:misha':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     gtrack.create_dirs</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="fl">60</span> <span class="op">*</span> <span class="fl">60</span><span class="op">)</span> <span class="co"># allow 2 hours for loading large files</span></span></code></pre></div>
<div class="section level3">
<h3 id="creating-a-misha-genome">Creating a <code>misha</code> genome<a class="anchor" aria-label="anchor" href="#creating-a-misha-genome"></a>
</h3>
<p>We will start by creating a misha database for mm10 genome. If you already have an <code>mm10</code> misha database you can skip this part and load the genome using <code>gsetroot("path/to/mm10")</code>.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/misha.ext/man/gdb.create_genome.html" class="external-link">gdb.create_genome</a></span><span class="op">(</span><span class="st">"mm10"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://tanaylab.github.io/misha/reference/gdb.init.html" class="external-link">gsetroot</a></span><span class="op">(</span><span class="st">"mm10"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="downloading-example-data">Downloading Example Data<a class="anchor" aria-label="anchor" href="#downloading-example-data"></a>
</h3>
<p>See <a href="#running-iceqream-on-your-own-data">below</a> for minimal instructions on how to use IceQream on your own data.</p>
<p>For this tutorial, we'll use pre-prepared data from a mouse gastrulation trajectory. Let's download and load this data:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://iceqream.s3.eu-west-1.amazonaws.com/gastrulation-example.tar.gz"</span>, <span class="st">"gastrulation-example.tar.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span><span class="st">"gastrulation-example.tar.gz"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">peak_intervals</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"gastrulation-example/peak_intervals.tsv"</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">atac_scores</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_rds.html" class="external-link">read_rds</a></span><span class="op">(</span><span class="st">"gastrulation-example/atac_scores.rds"</span><span class="op">)</span></span>
<span><span class="va">additional_features</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_rds.html" class="external-link">read_rds</a></span><span class="op">(</span><span class="st">"gastrulation-example/additional_features.rds"</span><span class="op">)</span></span>
<span><span class="va">normalization_intervals</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="st">"gastrulation-example/gastrulation_intervals.tsv"</span>, show_col_types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Let's examine the structure of our input data:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Peak intervals</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Peak Intervals:"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Peak Intervals:"</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">peak_intervals</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 6</span></span></span>
<span><span class="co">#&gt;   chrom   start     end peak_name   const tss_dist</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> chr1  3<span style="text-decoration: underline;">003</span>620 3<span style="text-decoration: underline;">003</span>920 1_Xkr4_3004 FALSE   <span style="text-decoration: underline;">211</span>711</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> chr1  3<span style="text-decoration: underline;">008</span>880 3<span style="text-decoration: underline;">009</span>180 1_Xkr4_3009 FALSE   <span style="text-decoration: underline;">206</span>451</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> chr1  3<span style="text-decoration: underline;">035</span>780 3<span style="text-decoration: underline;">036</span>080 1_Xkr4_3036 FALSE   <span style="text-decoration: underline;">179</span>551</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> chr1  3<span style="text-decoration: underline;">062</span>880 3<span style="text-decoration: underline;">063</span>180 1_Xkr4_3063 FALSE   <span style="text-decoration: underline;">152</span>451</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> chr1  3<span style="text-decoration: underline;">191</span>760 3<span style="text-decoration: underline;">192</span>060 1_Xkr4_3192 FALSE    <span style="text-decoration: underline;">23</span>571</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> chr1  3<span style="text-decoration: underline;">264</span>000 3<span style="text-decoration: underline;">264</span>300 1_Xkr4_3264 FALSE   -<span style="color: #BB0000; text-decoration: underline;">48</span><span style="color: #BB0000;">368</span></span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Number of peaks:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">peak_intervals</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of peaks: 99291"</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># ATAC scores</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"\nATAC Scores:"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "\nATAC Scores:"</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">atac_scores</span><span class="op">)</span></span>
<span><span class="co">#&gt;                    bin1        bin2       bin3       bin4</span></span>
<span><span class="co">#&gt; 1_Xkr4_3004 0.004287514 0.005028976 0.04591491 0.02115745</span></span>
<span><span class="co">#&gt; 1_Xkr4_3009 0.016835629 0.020535310 0.01645209 0.05152993</span></span>
<span><span class="co">#&gt; 1_Xkr4_3036 0.314234571 0.351552061 0.24488575 0.21536988</span></span>
<span><span class="co">#&gt; 1_Xkr4_3063 0.302141442 0.368683558 0.14807723 0.09834478</span></span>
<span><span class="co">#&gt; 1_Xkr4_3192 0.234917748 0.159842916 0.24823286 0.40749248</span></span>
<span><span class="co">#&gt; 1_Xkr4_3264 0.057764122 0.079473969 0.05351386 0.05018866</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Number of peaks:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">atac_scores</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of peaks: 99291"</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Number of bins:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">atac_scores</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of bins: 4"</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Additional features</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"\nAdditional Features:"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "\nAdditional Features:"</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">additional_features</span><span class="op">)</span></span>
<span><span class="co">#&gt;               cg_cont k4me3 k27me3 k27ac prox_bin1_punc_all spatial_ratio</span></span>
<span><span class="co">#&gt; 1_Xkr4_3004 0.6949901     0      0     0          0.2781718      4.764228</span></span>
<span><span class="co">#&gt; 1_Xkr4_3009 0.4440685     0      0     0          0.3025915      7.041941</span></span>
<span><span class="co">#&gt; 1_Xkr4_3036 1.0952665     0      0     0          0.2357028      8.936077</span></span>
<span><span class="co">#&gt; 1_Xkr4_3063 0.2247980     0      0     0          0.3126778      7.760096</span></span>
<span><span class="co">#&gt; 1_Xkr4_3192 0.9602237     0      0     0          0.6481828      5.660977</span></span>
<span><span class="co">#&gt; 1_Xkr4_3264 0.9239115     0      0     0          0.3089618      3.679446</span></span>
<span><span class="co">#&gt;                 TT       CT       GT       AT       TC       CC       GC</span></span>
<span><span class="co">#&gt; 1_Xkr4_3004 2.1875 8.181818 4.482759 2.500000 7.567568 4.615385 4.897959</span></span>
<span><span class="co">#&gt; 1_Xkr4_3009 5.0000 8.181818 3.103448 4.642857 6.216216 6.923077 4.285714</span></span>
<span><span class="co">#&gt; 1_Xkr4_3036 9.3750 4.090909 6.206897 4.642857 4.594595 2.820513 3.265306</span></span>
<span><span class="co">#&gt; 1_Xkr4_3063 4.3750 6.136364 5.517241 6.071429 7.297297 2.820513 2.653061</span></span>
<span><span class="co">#&gt; 1_Xkr4_3192 6.8750 6.363636 5.172414 8.571429 5.405405 4.358974 2.653061</span></span>
<span><span class="co">#&gt; 1_Xkr4_3264 7.5000 7.500000 4.827586 4.285714 7.297297 4.871795 3.877551</span></span>
<span><span class="co">#&gt;                   AC       TG        CG       GG       AG       TA       CA</span></span>
<span><span class="co">#&gt; 1_Xkr4_3004 3.103448 6.842105 0.9756098 7.435897 5.555556 1.111111 5.526316</span></span>
<span><span class="co">#&gt; 1_Xkr4_3009 7.241379 6.052632 0.4878049 2.051282 4.888889 4.444444 7.105263</span></span>
<span><span class="co">#&gt; 1_Xkr4_3036 4.482759 4.736842 1.2195122 6.153846 6.222222 4.814815 6.052632</span></span>
<span><span class="co">#&gt; 1_Xkr4_3063 5.172414 6.052632 0.0000000 4.102564 5.555556 4.074074 7.368421</span></span>
<span><span class="co">#&gt; 1_Xkr4_3192 4.827586 6.315789 0.9756098 2.820513 3.555556 8.518519 3.947368</span></span>
<span><span class="co">#&gt; 1_Xkr4_3264 5.517241 5.789474 1.4634146 2.564103 4.444444 3.703704 6.052632</span></span>
<span><span class="co">#&gt;                   GA     AA</span></span>
<span><span class="co">#&gt; 1_Xkr4_3004 4.594595 2.8125</span></span>
<span><span class="co">#&gt; 1_Xkr4_3009 4.594595 2.8125</span></span>
<span><span class="co">#&gt; 1_Xkr4_3036 4.594595 4.6875</span></span>
<span><span class="co">#&gt; 1_Xkr4_3063 5.135135 7.1875</span></span>
<span><span class="co">#&gt; 1_Xkr4_3192 4.054054 6.2500</span></span>
<span><span class="co">#&gt; 1_Xkr4_3264 3.783784 3.4375</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Number of peaks:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">additional_features</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of peaks: 99291"</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Number of features:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">additional_features</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Number of features: 22"</span></span></code></pre></div>
<p>The <code>peak_intervals</code> dataframe contains the genomic positions of accessibility peaks. The <code>atac_scores</code> matrix contains ATAC-seq signal intensities for each peak across different stages of the trajectory. <code>additional_features</code> includes extra genomic features for each peak.</p>
</div>
</div>
<div class="section level2">
<h2 id="computing-motif-energies">Computing Motif Energies<a class="anchor" aria-label="anchor" href="#computing-motif-energies"></a>
</h2>
<p>The first step in the IceQream pipeline is to compute motif energies for each transcription factor model and each peak. This process is computationally intensive, as it calculates energies for over 20,000 motifs from various databases.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">motif_energies</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_motif_energies.html">compute_motif_energies</a></span><span class="op">(</span><span class="va">peak_intervals</span>, <span class="va">motif_db</span>, normalization_intervals <span class="op">=</span> <span class="va">normalization_intervals</span><span class="op">)</span></span></code></pre></div>
<p>For this tutorial, we'll use pre-computed motif energies:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://iceqream.s3.eu-west-1.amazonaws.com/gastrulation_energies.tar.gz"</span>, <span class="st">"gastrulation_energies.tar.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span><span class="st">"gastrulation_energies.tar.gz"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">motif_energies</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_rds.html" class="external-link">read_rds</a></span><span class="op">(</span><span class="st">"gastrulation_energies.rds"</span><span class="op">)</span></span>
<span><span class="va">motif_energies</span> <span class="op">&lt;-</span> <span class="va">motif_energies</span><span class="op">[</span><span class="va">peak_intervals</span><span class="op">$</span><span class="va">peak_name</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Motif energy matrix dimensions:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">motif_energies</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">" x "</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Motif energy matrix dimensions: 99291 x 21862"</span></span></code></pre></div>
<p>For a less memory and computationally intensive analysis, you can use the SCENIC motif clusters instead of all motifs (See <a href="#performance-considerations">Performance Considerations</a> section).</p>
</div>
<div class="section level2">
<h2 id="running-iceqream">Running IceQream<a class="anchor" aria-label="anchor" href="#running-iceqream"></a>
</h2>
<p>Now we're ready to run the IceQream regression:</p>
<p>This takes ~25 minutes on a 40 core machine with 256GB of RAM.</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traj_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/iq_regression.html">iq_regression</a></span><span class="op">(</span></span>
<span>    peak_intervals <span class="op">=</span> <span class="va">peak_intervals</span>,</span>
<span>    atac_scores <span class="op">=</span> <span class="va">atac_scores</span>,</span>
<span>    motif_energies <span class="op">=</span> <span class="va">motif_energies</span>,</span>
<span>    additional_features <span class="op">=</span> <span class="va">additional_features</span>,</span>
<span>    norm_intervals <span class="op">=</span> <span class="va">normalization_intervals</span>,</span>
<span>    seed <span class="op">=</span> <span class="fl">60427</span>,</span>
<span>    frac_train <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>    max_motif_num <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Training on <span style="color: #0000BB;">79432</span> intervals (80%) and testing on <span style="color: #0000BB;">19859</span> intervals (20%)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Number of peaks: <span style="color: #0000BB;">79432</span></span></span>
<span><span class="co">#&gt; → Extracting sequences...</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Calculating correlations between <span style="color: #0000BB;">21862</span> motif energies and ATAC difference...</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Selected <span style="color: #0000BB;">1558</span> (out of <span style="color: #0000BB;">21862</span>) features with absolute correlation &gt;= <span style="color: #0000BB;">0.05</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Running first round of regression, # of features: <span style="color: #0000BB;">1558</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning: glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Taking <span style="color: #0000BB;">743</span> features with beta &gt;= <span style="color: #0000BB;">0.003</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Running second round of regression...</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Clustering <span style="color: #0000BB;">740</span> features into <span style="color: #0000BB;">30</span> clusters...</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Choosing top <span style="color: #0000BB;">30</span> features clusters</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Features left: <span style="color: #0000BB;">740</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Learning a model for each motif cluster...</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Infering energies...</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Finished running model. Number of non-zero coefficients: <span style="color: #0000BB;">201</span> (out of <span style="color: #0000BB;">208</span>). R^2: <span style="color: #0000BB;">0.309612868425167</span></span></span>
<span><span class="co">#&gt; → Filtering the model</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Using <span style="color: #0000BB;">7940</span> samples for filtering</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> R^2 of the full model: <span style="color: #0000BB;">0.29602387224171</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Filtering features with R^2 &lt; <span style="color: #0000BB;">5e-04</span> and bits &lt; <span style="color: #0000BB;">1.75</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">JASPAR.nub</span> (<span style="font-weight: bold;">----TTAGCAT----</span>): <span style="color: #0000BB;">0.0167275374369183</span>. Bits: <span style="color: #0000BB;">7.00354991194335</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.cisbp__M08572</span> (<span style="font-weight: bold;">---T-----------</span>): <span style="color: #0000BB;">0.0026997995326517</span>. Bits: <span style="color: #0000BB;">1.10663727169756</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.predrem__nrMotif1342</span> (<span style="font-weight: bold;">---G-CAAATGG---</span>): <span style="color: #0000BB;">0.00634826224842949</span>. Bits: <span style="color: #0000BB;">7.99024178403719</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.cisbp__M00835</span> (<span style="font-weight: bold;">----GTGT-A-----</span>): <span style="color: #0000BB;">0.0305228554917784</span>. Bits: <span style="color: #0000BB;">4.57075101707182</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.hdpi__CREB1</span> (<span style="font-weight: bold;">----TGAC--A----</span>): <span style="color: #0000BB;">7.75643877909959e-05</span>. Bits: <span style="color: #0000BB;">1.89464022573617</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.homer__ACCATCTGTT_NeuroG2</span> (<span style="font-weight: bold;">-----CAGATGG---</span>): <span style="color: #0000BB;">-0.000120197791811605</span>. Bits: <span style="color: #0000BB;">4.59034584610514</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.swissregulon__hs__ID4</span> (<span style="font-weight: bold;">-----GCACCTG---</span>): <span style="color: #0000BB;">0.00278910362030049</span>. Bits: <span style="color: #0000BB;">10.7708040996155</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.taipale_cyt_meth__ESX1_YYAATTAN_eDBD</span> (<span style="font-weight: bold;">------TAAT-G---</span>): <span style="color: #0000BB;">0.0173484629685431</span>. Bits: <span style="color: #0000BB;">3.82249971740051</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.jaspar__MA1621.1</span> (<span style="font-weight: bold;">---CACCTGC-----</span>): <span style="color: #0000BB;">0.000955852850026528</span>. Bits: <span style="color: #0000BB;">9.97745953946352</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">JOLMA.FIGLA_di_DBD</span> (<span style="font-weight: bold;">---ACAGGT------</span>): <span style="color: #0000BB;">0.000527737749653978</span>. Bits: <span style="color: #0000BB;">5.74258158770429</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.cisbp__M01516</span> (<span style="font-weight: bold;">----ATGCA------</span>): <span style="color: #0000BB;">4.75776194436284e-06</span>. Bits: <span style="color: #0000BB;">3.33918513257243</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.taipale_cyt_meth__TFAP4_ANCATATGNT_eDBD_meth</span> (<span style="font-weight: bold;">---ACCAAATGTT--</span>): <span style="color: #0000BB;">0.000743126659866977</span>. Bits: <span style="color: #0000BB;">9.24966897852248</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.swissregulon__mm__Gata1</span> (<span style="font-weight: bold;">-----GATA------</span>): <span style="color: #0000BB;">0.0059904574972684</span>. Bits: <span style="color: #0000BB;">3.58604128728144</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.taipale_cyt_meth__TGIF2_NTGACAGN_FL</span> (<span style="font-weight: bold;">----CTGACA-----</span>): <span style="color: #0000BB;">0.000589606574210977</span>. Bits: <span style="color: #0000BB;">5.47428832011949</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">HOCOMOCO.SUH_MOUSE.H11MO.0.A</span> (<span style="font-weight: bold;">-----TGGGA-----</span>): <span style="color: #0000BB;">0.000956012692981068</span>. Bits: <span style="color: #0000BB;">2.60135226413422</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.jaspar__MA0869.2</span> (<span style="font-weight: bold;">----AAAT--A----</span>): <span style="color: #0000BB;">0.00149994976970297</span>. Bits: <span style="color: #0000BB;">2.43576636244517</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.flyfactorsurvey__HLHm5_FlyReg_FBgn0002631</span> (<span style="font-weight: bold;">------A--------</span>): <span style="color: #0000BB;">0.00010975195439078</span>. Bits: <span style="color: #0000BB;">0.788368839986324</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">HOMER.PQM_1</span> (<span style="font-weight: bold;">----TGATAAG----</span>): <span style="color: #0000BB;">0.00117074463927358</span>. Bits: <span style="color: #0000BB;">5.35487254467996</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.hocomoco__SMAD4_MOUSE.H11MO.0.A</span> (<span style="font-weight: bold;">---GTG-CAGAC---</span>): <span style="color: #0000BB;">0.00219652282477695</span>. Bits: <span style="color: #0000BB;">7.85384961351908</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.cisbp__M02320</span> (<span style="font-weight: bold;">-----GAT-------</span>): <span style="color: #0000BB;">0.00112380163185027</span>. Bits: <span style="color: #0000BB;">3.4817723117682</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.hocomoco__ZIC2_MOUSE.H11MO.0.C</span> (<span style="font-weight: bold;">-------CAGGTG--</span>): <span style="color: #0000BB;">0.00484774757259671</span>. Bits: <span style="color: #0000BB;">7.66340237394876</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">JASPAR.SOX14</span> (<span style="font-weight: bold;">----ACAATG-----</span>): <span style="color: #0000BB;">0.020314669301463</span>. Bits: <span style="color: #0000BB;">6.01846819360433</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.taipale_cyt_meth__TCF7L1_ASATCAAAS_eDBD_meth_repr</span> (<span style="font-weight: bold;">----CATCAAAG---</span>): <span style="color: #0000BB;">0.0102889730006238</span>. Bits: <span style="color: #0000BB;">5.82383823306701</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.nitta__SoxN_TGAGGT40NTCC_KK_NCGAACAATNN_m1_c2</span> (<span style="font-weight: bold;">----CGTAC------</span>): <span style="color: #0000BB;">0.00015623097637113</span>. Bits: <span style="color: #0000BB;">2.83138146510066</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.dbtfbs__mm_MYOG_myocyte_ENCSR000AIW_merged_N1</span> (<span style="font-weight: bold;">--AACAGCTG-----</span>): <span style="color: #0000BB;">0.00164317632907796</span>. Bits: <span style="color: #0000BB;">6.33133125285195</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">HOMER.TATA_box_1</span> (<span style="font-weight: bold;">---CTATAAAAG---</span>): <span style="color: #0000BB;">-3.49324694661757e-05</span>. Bits: <span style="color: #0000BB;">6.0079726304727</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.predrem__nrMotif1713</span> (<span style="font-weight: bold;">----GTCACA-----</span>): <span style="color: #0000BB;">0.000223555700244238</span>. Bits: <span style="color: #0000BB;">5.74538254268905</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.transfac_public__M00075</span> (<span style="font-weight: bold;">--TCACCCCAC----</span>): <span style="color: #0000BB;">-0.00041752947580298</span>. Bits: <span style="color: #0000BB;">3.92211592781422</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.taipale_cyt_meth__HMX1_NCAMTTAAN_eDBD_meth_repr</span> (<span style="font-weight: bold;">-----AAGTG-----</span>): <span style="color: #0000BB;">-0.00057932247059217</span>. Bits: <span style="color: #0000BB;">6.61434403320222</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → R^2 added by <span style="color: #00BB00;">SCENIC.taipale_tf_pairs__HOXB2_PITX1_TAATKRNGGATTA_CAP_repr</span> (<span style="font-weight: bold;">TAATTA-GGATTA</span>): <span style="color: #0000BB;">0.000496812804951952</span>. Bits: <span style="color: #0000BB;">10.8696138190304</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Removing the following features with bits &lt; <span style="color: #0000BB;">1.75</span>: <span style="color: #0000BB;">"SCENIC.cisbp__M08572"</span> and <span style="color: #0000BB;">"SCENIC.flyfactorsurvey__HLHm5_FlyReg_FBgn0002631"</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Trying to remove the following features with R^2 &lt; <span style="color: #0000BB;">5e-04</span>: <span style="color: #0000BB;">"SCENIC.hdpi__CREB1"</span>, <span style="color: #0000BB;">"SCENIC.homer__ACCATCTGTT_NeuroG2"</span>, <span style="color: #0000BB;">"SCENIC.cisbp__M01516"</span>, <span style="color: #0000BB;">"SCENIC.flyfactorsurvey__HLHm5_FlyReg_FBgn0002631"</span>, <span style="color: #0000BB;">"SCENIC.nitta__SoxN_TGAGGT40NTCC_KK_NCGAACAATNN_m1_c2"</span>, <span style="color: #0000BB;">"HOMER.TATA_box_1"</span>, <span style="color: #0000BB;">"SCENIC.predrem__nrMotif1713"</span>, <span style="color: #0000BB;">"SCENIC.transfac_public__M00075"</span>, <span style="color: #0000BB;">"SCENIC.taipale_cyt_meth__HMX1_NCAMTTAAN_eDBD_meth_repr"</span>, and <span style="color: #0000BB;">"SCENIC.taipale_tf_pairs__HOXB2_PITX1_TAATKRNGGATTA_CAP_repr"</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning: glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → Removing <span style="color: #0000BB;">"SCENIC.transfac_public__M00075"</span> changed the R^2 by <span style="color: #0000BB;">-0.000411123637068522</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → Removing <span style="color: #0000BB;">"SCENIC.homer__ACCATCTGTT_NeuroG2"</span> changed the R^2 by <span style="color: #0000BB;">-6.44393580477853e-05</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → <span style="color: #00BB00;">Not removing</span> <span style="color: #0000BB;">"HOMER.TATA_box_1"</span> (changed the R^2 by only <span style="color: #0000BB;">0.000758037995756855</span>).</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → Removing <span style="color: #0000BB;">"SCENIC.cisbp__M01516"</span> changed the R^2 by <span style="color: #0000BB;">-2.65246572077604e-05</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → Removing <span style="color: #0000BB;">"SCENIC.hdpi__CREB1"</span> changed the R^2 by <span style="color: #0000BB;">-0.000189532430595718</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → Removing <span style="color: #0000BB;">"SCENIC.nitta__SoxN_TGAGGT40NTCC_KK_NCGAACAATNN_m1_c2"</span> changed the R^2 by <span style="color: #0000BB;">0.000190831533410263</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → Removing <span style="color: #0000BB;">"SCENIC.predrem__nrMotif1713"</span> changed the R^2 by <span style="color: #0000BB;">0.000233902642093553</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; → <span style="color: #00BB00;">Not removing</span> <span style="color: #0000BB;">"SCENIC.taipale_tf_pairs__HOXB2_PITX1_TAATKRNGGATTA_CAP_repr"</span> (changed the R^2 by only <span style="color: #0000BB;">0.00053681107061454</span>).</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Removed <span style="color: #0000BB;">2</span> features with bits &lt; <span style="color: #0000BB;">1.75</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Removed <span style="color: #0000BB;">7</span> features with R^2 &lt; <span style="color: #0000BB;">5e-04</span></span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): glmnet.fit: algorithm did not converge</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> After filtering: Number of non-zero coefficients: <span style="color: #0000BB;">166</span> (out of <span style="color: #0000BB;">172</span>). R^2: <span style="color: #0000BB;">0.304057017984406</span>. Number of models: <span style="color: #0000BB;">21</span></span></span>
<span><span class="co">#&gt; → Infering trajectory motifs on the test set</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Extracting sequences...</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Computing motif energies for <span style="color: #0000BB;">19859</span> intervals</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Finished IQ regression.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of motifs: <span style="color: #0000BB;">21</span></span></span>
<span><span class="co">#&gt; R^2 train: <span style="color: #0000BB;">0.304</span></span></span>
<span><span class="co">#&gt; R^2 test: <span style="color: #0000BB;">0.305</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Run `plot_traj_model_report(traj_model)` to visualize the model features</span></span>
<span><span class="co">#&gt; Run `plot_prediction_scatter(traj_model)` to visualize the model predictions</span></span></code></pre></div>
<p>Let's examine the output:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">traj_model</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">&lt;TrajectoryModel&gt;</span> with <span style="color: #0000BB;">21</span> motifs and <span style="color: #0000BB;">22</span> additional features</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Slots include:</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@model</span>: A GLM model object. Number of non-zero coefficients: <span style="color: #0000BB;">166</span></span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@motif_models</span>: A named list of motif models. Each element contains PSSM and</span></span>
<span><span class="co">#&gt; spatial model (<span style="color: #0000BB;">21</span> models: <span style="color: #0000BB;">"JASPAR.nub"</span>, <span style="color: #0000BB;">"SCENIC.predrem__nrMotif1342"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"SCENIC.cisbp__M00835"</span>, <span style="color: #0000BB;">"SCENIC.swissregulon__hs__ID4"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"SCENIC.taipale_cyt_meth__ESX1_YYAATTAN_eDBD"</span>, <span style="color: #0000BB;">"SCENIC.jaspar__MA1621.1"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"JOLMA.FIGLA_di_DBD"</span>, <span style="color: #0000BB;">"SCENIC.taipale_cyt_meth__TFAP4_ANCATATGNT_eDBD_meth"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"SCENIC.swissregulon__mm__Gata1"</span>, <span style="color: #0000BB;">"SCENIC.taipale_cyt_meth__TGIF2_NTGACAGN_FL"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"HOCOMOCO.SUH_MOUSE.H11MO.0.A"</span>, <span style="color: #0000BB;">"SCENIC.jaspar__MA0869.2"</span>, <span style="color: #0000BB;">"HOMER.PQM_1"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"SCENIC.hocomoco__SMAD4_MOUSE.H11MO.0.A"</span>, <span style="color: #0000BB;">"SCENIC.cisbp__M02320"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"SCENIC.hocomoco__ZIC2_MOUSE.H11MO.0.C"</span>, <span style="color: #0000BB;">"JASPAR.SOX14"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"SCENIC.taipale_cyt_meth__TCF7L1_ASATCAAAS_eDBD_meth_repr"</span>, …,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"HOMER.TATA_box_1"</span>, and</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"SCENIC.taipale_tf_pairs__HOXB2_PITX1_TAATKRNGGATTA_CAP_repr"</span>)</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@additional_features</span>: A data frame of additional features (<span style="color: #0000BB;">22</span> features)</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@coefs</span>: A data frame of coefficients (<span style="color: #0000BB;">43</span> elements)</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@model_features</span>: A matrix of the model features (logistic functions of the</span></span>
<span><span class="co">#&gt; motif models energies, dimensions: <span style="color: #0000BB;">99291</span>x<span style="color: #0000BB;">172</span>)</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@normalized_energies</span>: A matrix of normalized energies of the model features</span></span>
<span><span class="co">#&gt; (dimensions: <span style="color: #0000BB;">99291</span>x<span style="color: #0000BB;">21</span>)</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@type</span>: A vector the length of the number of peaks, indicating whether each</span></span>
<span><span class="co">#&gt; peak is a training ('train') or a prediction peak ('test')</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@diff_score</span>: A numeric value representing the difference score the model was</span></span>
<span><span class="co">#&gt; trained on (<span style="color: #0000BB;">99291</span> elements)</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@predicted_diff_score</span>: A numeric value representing the predicted difference</span></span>
<span><span class="co">#&gt; score</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@initial_prego_models</span>: A list of prego models used in the initial phase of</span></span>
<span><span class="co">#&gt; the algorithm (<span style="color: #0000BB;">0</span> models)</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@peak_intervals</span>: A data frame containing the peak intervals (<span style="color: #0000BB;">99291</span> elements)</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@normalization_intervals</span>: A data frame containing the intervals used for</span></span>
<span><span class="co">#&gt; energy normalization (<span style="color: #0000BB;">129963</span> elements)</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@features_r2</span>: A numeric vector of the added R^2 values for each feature (<span style="color: #0000BB;">21</span></span></span>
<span><span class="co">#&gt; elements)</span></span>
<span><span class="co">#&gt; • <span style="color: #00BB00;">@params</span>: A list of parameters used for training (including:</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"energy_norm_quantile"</span>, <span style="color: #0000BB;">"norm_energy_max"</span>, <span style="color: #0000BB;">"min_energy"</span>, <span style="color: #0000BB;">"alpha"</span>, <span style="color: #0000BB;">"lambda"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"peaks_size"</span>, <span style="color: #0000BB;">"spat_num_bins"</span>, <span style="color: #0000BB;">"spat_bin_size"</span>, <span style="color: #0000BB;">"distilled_features"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"n_clust_factor"</span>, <span style="color: #0000BB;">"seed"</span>, <span style="color: #0000BB;">"features_bits"</span>, <span style="color: #0000BB;">"r2_threshold"</span>, <span style="color: #0000BB;">"bits_threshold"</span>,</span></span>
<span><span class="co">#&gt; and <span style="color: #0000BB;">"stats"</span>)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; R^2 train: <span style="color: #0000BB;">0.304</span></span></span>
<span><span class="co">#&gt; R^2 test: <span style="color: #0000BB;">0.305</span></span></span></code></pre></div>
<p>The <code>TrajectoryModel</code> object contains components such as the regression model, motif models, and predicted accessibility scores.</p>
</div>
<div class="section level2">
<h2 id="visualizing-results">Visualizing Results<a class="anchor" aria-label="anchor" href="#visualizing-results"></a>
</h2>
<p>Let's start with a scatter plot of observed vs. predicted accessibility changes:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_prediction_scatter.html">plot_prediction_scatter</a></span><span class="op">(</span><span class="va">traj_model</span><span class="op">)</span></span></code></pre></div>
<p><img src="iceqream_files/figure-html/prediction_scatter-1.png" width="672"></p>
<p>This plot shows how well our model predicts accessibility changes. Points closer to the diagonal line indicate better predictions. We measure the accuracy of the model using the coefficient of determination (R^2).</p>
<div class="section level3">
<h3 id="model-report">Model report<a class="anchor" aria-label="anchor" href="#model-report"></a>
</h3>
<p>Next, let's look at the model report, which provides detailed information about the motifs and their contributions:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_traj_model_report.html">plot_traj_model_report</a></span><span class="op">(</span><span class="va">traj_model</span>, filename <span class="op">=</span> <span class="st">"model_report.pdf"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"model_report.pdf"</span><span class="op">)</span></span></code></pre></div>
<p><embed src="model_report.pdf"></embed><!-- --></p>
</div>
<div class="section level3">
<h3 id="interpreting-the-trajectory-model-report">Interpreting the trajectory model report<a class="anchor" aria-label="anchor" href="#interpreting-the-trajectory-model-report"></a>
</h3>
<p>The model report provides several key pieces of information (from left to right):</p>
<ol style="list-style-type: decimal">
<li>Motif logos show the inferred sequence preferences for each transcription factor model.</li>
<li>Response curves show how the accessibility changes as a function of binding energy for each TF.</li>
<li>Barplots show the coefficient of each non-linear term of every motif in the model.</li>
<li>Spatial model curves show the parameters of the spatial model for each TF. The R² values indicate the predictive power each TF adds to the model.</li>
<li>Spatial curves show the frequencey of each TF binding site along the peaks from the bottom 10% (blue) and top 10% (red) of the differential accessibility (dAP) distribution.</li>
<li>Boxplots show the distribution of ATAC differences (dAP, y-axis) for bins of binding energy (x-axis) for each TF.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="renaming-the-motif-models">Renaming the motif models<a class="anchor" aria-label="anchor" href="#renaming-the-motif-models"></a>
</h2>
<p>You can rename the motif models to more informative names, either manually using <code>rename_motif_models</code> or automatically using <code>match_traj_model_motif_names</code>:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">names_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/match_traj_model_motif_names.html">match_traj_model_motif_names</a></span><span class="op">(</span><span class="va">traj_model</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Matching motif names, note that this might take a while.</span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">JASPAR.nub</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Oct4"</span>, PSSM correlation = <span style="color: #0000BB;">0.898696170118534</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.predrem__nrMotif1342</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Atoh1"</span>, PSSM correlation = <span style="color: #0000BB;">0.74898042552435</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.cisbp__M00835</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Tbx5"</span>, PSSM correlation = <span style="color: #0000BB;">0.904736231501015</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.swissregulon__hs__ID4</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.E2A_1"</span>, PSSM correlation = <span style="color: #0000BB;">0.867727837996119</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.taipale_cyt_meth__ESX1_YYAATTAN_eDBD</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Isl1"</span>, PSSM correlation = <span style="color: #0000BB;">0.869760207503598</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.jaspar__MA1621.1</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.E2A_1"</span>, PSSM correlation = <span style="color: #0000BB;">0.918354413557226</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">JOLMA.FIGLA_di_DBD</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.MyoG"</span>, PSSM correlation = <span style="color: #0000BB;">0.721047792856512</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.taipale_cyt_meth__TFAP4_ANCATATGNT_eDBD_meth</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Atoh1"</span>, PSSM correlation = <span style="color: #0000BB;">0.669100601091648</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.swissregulon__mm__Gata1</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.GATA3_2"</span>, PSSM correlation = <span style="color: #0000BB;">0.95177409403423</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.taipale_cyt_meth__TGIF2_NTGACAGN_FL</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Meis1"</span>, PSSM correlation = <span style="color: #0000BB;">0.733098848365572</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">HOCOMOCO.SUH_MOUSE.H11MO.0.A</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.E2F6"</span>, PSSM correlation = <span style="color: #0000BB;">0.657723158470179</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.jaspar__MA0869.2</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Pit1"</span>, PSSM correlation = <span style="color: #0000BB;">0.532601141276059</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">HOMER.PQM_1</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.PQM_1"</span>, PSSM correlation = <span style="color: #0000BB;">0.94889261728701</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.hocomoco__SMAD4_MOUSE.H11MO.0.A</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.TboxSmad"</span>, PSSM correlation = <span style="color: #0000BB;">0.971715050385645</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.cisbp__M02320</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.GATA3_2"</span>, PSSM correlation = <span style="color: #0000BB;">0.802823387218573</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.hocomoco__ZIC2_MOUSE.H11MO.0.C</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.E_box_2"</span>, PSSM correlation = <span style="color: #0000BB;">0.710572263191446</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">JASPAR.SOX14</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Tcf3"</span>, PSSM correlation = <span style="color: #0000BB;">0.67164203806063</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.taipale_cyt_meth__TCF7L1_ASATCAAAS_eDBD_meth_repr</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Tcf4"</span>, PSSM correlation = <span style="color: #0000BB;">0.950211584477428</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.dbtfbs__mm_MYOG_myocyte_ENCSR000AIW_merged_N1</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Atoh1"</span>, PSSM correlation = <span style="color: #0000BB;">0.886109683951447</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">HOMER.TATA_box_1</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.TATA_box"</span>, PSSM correlation = <span style="color: #0000BB;">0.867728627783548</span></span></span>
<span><span class="co">#&gt; → Matching <span style="color: #00BB00;">SCENIC.taipale_tf_pairs__HOXB2_PITX1_TAATKRNGGATTA_CAP_repr</span></span></span>
<span><span class="co">#&gt; → Matched with <span style="color: #0000BB;">"HOMER.Lhx2"</span>, PSSM correlation = <span style="color: #0000BB;">0.525391618049386</span></span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">names_map</span></span>
<span><span class="co">#&gt;                                HOCOMOCO.SUH_MOUSE.H11MO.0.A </span></span>
<span><span class="co">#&gt;                                                      "E2F6" </span></span>
<span><span class="co">#&gt;                                                 HOMER.PQM_1 </span></span>
<span><span class="co">#&gt;                                                     "PQM_1" </span></span>
<span><span class="co">#&gt;                                            HOMER.TATA_box_1 </span></span>
<span><span class="co">#&gt;                                                  "TATA_box" </span></span>
<span><span class="co">#&gt;                                                JASPAR.SOX14 </span></span>
<span><span class="co">#&gt;                                                      "Tcf3" </span></span>
<span><span class="co">#&gt;                                                  JASPAR.nub </span></span>
<span><span class="co">#&gt;                                                      "Oct4" </span></span>
<span><span class="co">#&gt;                                          JOLMA.FIGLA_di_DBD </span></span>
<span><span class="co">#&gt;                                                      "MyoG" </span></span>
<span><span class="co">#&gt;                                        SCENIC.cisbp__M00835 </span></span>
<span><span class="co">#&gt;                                                      "Tbx5" </span></span>
<span><span class="co">#&gt;                                        SCENIC.cisbp__M02320 </span></span>
<span><span class="co">#&gt;                                                   "GATA3_2" </span></span>
<span><span class="co">#&gt;        SCENIC.dbtfbs__mm_MYOG_myocyte_ENCSR000AIW_merged_N1 </span></span>
<span><span class="co">#&gt;                                                     "Atoh1" </span></span>
<span><span class="co">#&gt;                      SCENIC.hocomoco__SMAD4_MOUSE.H11MO.0.A </span></span>
<span><span class="co">#&gt;                                                  "TboxSmad" </span></span>
<span><span class="co">#&gt;                       SCENIC.hocomoco__ZIC2_MOUSE.H11MO.0.C </span></span>
<span><span class="co">#&gt;                                                   "E_box_2" </span></span>
<span><span class="co">#&gt;                                     SCENIC.jaspar__MA0869.2 </span></span>
<span><span class="co">#&gt;                                                      "Pit1" </span></span>
<span><span class="co">#&gt;                                     SCENIC.jaspar__MA1621.1 </span></span>
<span><span class="co">#&gt;                                                     "E2A_1" </span></span>
<span><span class="co">#&gt;                                 SCENIC.predrem__nrMotif1342 </span></span>
<span><span class="co">#&gt;                                                   "Atoh1.2" </span></span>
<span><span class="co">#&gt;                                SCENIC.swissregulon__hs__ID4 </span></span>
<span><span class="co">#&gt;                                                   "E2A_1.2" </span></span>
<span><span class="co">#&gt;                              SCENIC.swissregulon__mm__Gata1 </span></span>
<span><span class="co">#&gt;                                                 "GATA3_2.2" </span></span>
<span><span class="co">#&gt;                 SCENIC.taipale_cyt_meth__ESX1_YYAATTAN_eDBD </span></span>
<span><span class="co">#&gt;                                                      "Isl1" </span></span>
<span><span class="co">#&gt;    SCENIC.taipale_cyt_meth__TCF7L1_ASATCAAAS_eDBD_meth_repr </span></span>
<span><span class="co">#&gt;                                                      "Tcf4" </span></span>
<span><span class="co">#&gt;         SCENIC.taipale_cyt_meth__TFAP4_ANCATATGNT_eDBD_meth </span></span>
<span><span class="co">#&gt;                                                   "Atoh1.3" </span></span>
<span><span class="co">#&gt;                  SCENIC.taipale_cyt_meth__TGIF2_NTGACAGN_FL </span></span>
<span><span class="co">#&gt;                                                     "Meis1" </span></span>
<span><span class="co">#&gt; SCENIC.taipale_tf_pairs__HOXB2_PITX1_TAATKRNGGATTA_CAP_repr </span></span>
<span><span class="co">#&gt;                                                      "Lhx2"</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">traj_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rename_motif_models.html">rename_motif_models</a></span><span class="op">(</span><span class="va">traj_model</span>, <span class="va">names_map</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in eval(family$initialize): non-integer #successes in a binomial glm!</span></span>
<span><span class="co">#&gt; Warning: glmnet.fit: algorithm did not converge</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="exporting-the-model">Exporting the model<a class="anchor" aria-label="anchor" href="#exporting-the-model"></a>
</h2>
<p>You can export the minimal model representation to a list of PBM in order to use infer it on new data:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbm_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/traj_model_to_pbm_list.html">traj_model_to_pbm_list</a></span><span class="op">(</span><span class="va">traj_model</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Computing motif energies for <span style="color: #0000BB;">21</span> motifs on <span style="color: #0000BB;">129963</span> normalization intervals</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbm_list</span></span>
<span><span class="co">#&gt; $Oct4</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Oct4"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-22.8997500001511</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Atoh1.2</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Atoh1.2"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-22.7175598349943</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Tbx5</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Tbx5"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-23.7724491326609</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $E2A_1.2</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"E2A_1.2"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-21.2037809248507</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Isl1</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Isl1"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-25.5471422344978</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $E2A_1</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"E2A_1"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-21.4483677716017</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $MyoG</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"MyoG"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-23.3601778254574</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Atoh1.3</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Atoh1.3"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-22.1394656075513</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $GATA3_2.2</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"GATA3_2.2"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-25.6915278612726</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Meis1</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Meis1"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-24.7345338401995</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $E2F6</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"E2F6"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-25.9229999989994</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Pit1</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Pit1"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-26.0310794002464</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $PQM_1</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"PQM_1"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-24.6383171142784</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $TboxSmad</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"TboxSmad"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-22.3027239225011</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $GATA3_2</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"GATA3_2"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-26.5521923358056</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $E_box_2</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"E_box_2"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-22.6841209042217</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Tcf3</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Tcf3"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-23.5664496893186</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Tcf4</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Tcf4"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-23.5260544042839</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Atoh1</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Atoh1"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-22.55867400696</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $TATA_box</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"TATA_box"</span> with <span style="color: #0000BB;">15</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-23.5020346193338</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Lhx2</span></span>
<span><span class="co">#&gt; a <span style="color: #0000BB;">&lt;PBM&gt;</span> object named <span style="color: #0000BB;">"Lhx2"</span> with <span style="color: #0000BB;">13</span> positions (`@pssm`)</span></span>
<span><span class="co">#&gt; Energy normalization max = <span style="color: #0000BB;">-14.2673450054441</span> (`@max_energy`)</span></span>
<span><span class="co">#&gt; Spatial distribution with <span style="color: #0000BB;">149</span> spatial factors, from position <span style="color: #0000BB;">1</span> to <span style="color: #0000BB;">299</span> (<span style="color: #0000BB;">298</span> bp)</span></span>
<span><span class="co">#&gt; (`@spat`)</span></span>
<span><span class="co">#&gt; Includes a model with <span style="color: #0000BB;">4</span> coefficients (<span style="color: #0000BB;">"high-energy"</span>, <span style="color: #0000BB;">"higher-energy"</span>,</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">"low-energy"</span>, and <span style="color: #0000BB;">"sigmoid"</span>) (`@coefs`)</span></span></code></pre></div>
<p>You can now use <code>pbm_list.compute</code> or <code>pbm_list.gextract</code> to infer the model on new data:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new_intervals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    chrom <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>    start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3671720</span>, <span class="fl">4412460</span>, <span class="fl">4493400</span><span class="op">)</span>,</span>
<span>    end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3672020</span>, <span class="fl">4412760</span>, <span class="fl">4493700</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/pbm_list.gextract.html">pbm_list.gextract</a></span><span class="op">(</span><span class="va">pbm_list</span>, <span class="va">new_intervals</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Computing energies for <span style="color: #0000BB;">21</span> PBMs on <span style="color: #0000BB;">3</span> sequences</span></span>
<span><span class="co">#&gt;   chrom   start     end     Oct4  Atoh1.2     Tbx5 E2A_1.2     Isl1 E2A_1</span></span>
<span><span class="co">#&gt; 1  chr1 3671720 3672020 0.000000 0.000000 3.360820       0 3.729907     0</span></span>
<span><span class="co">#&gt; 2  chr1 4412460 4412760 2.438457 2.731101 3.491086       0 4.280369     0</span></span>
<span><span class="co">#&gt; 3  chr1 4493400 4493700 0.000000 0.000000 2.669436       0 3.894363     0</span></span>
<span><span class="co">#&gt;        MyoG Atoh1.3 GATA3_2.2    Meis1     E2F6     Pit1    PQM_1  TboxSmad</span></span>
<span><span class="co">#&gt; 1 2.0862625       0  3.806423 3.154064 4.018626 2.305464 1.271002 0.1399182</span></span>
<span><span class="co">#&gt; 2 2.3392323       0  5.520942 6.555252 5.438935 4.640641 5.010974 4.3378987</span></span>
<span><span class="co">#&gt; 3 0.5483327       0  4.763735 4.778387 3.007386 4.004477 1.685282 0.4491135</span></span>
<span><span class="co">#&gt;    GATA3_2   E_box_2     Tcf3      Tcf4    Atoh1 TATA_box Lhx2</span></span>
<span><span class="co">#&gt; 1 6.282286 1.2634071 0.000000 0.4460041 4.839001 0.000000    0</span></span>
<span><span class="co">#&gt; 2 4.587564 0.4109196 2.038721 6.6088638 0.000000 1.872493    0</span></span>
<span><span class="co">#&gt; 3 5.887933 0.9321862 2.091852 0.0000000 0.000000 2.445689    0</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># directly compute on sequences</span></span>
<span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="fu">prego</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/prego/man/intervals_to_seq.html" class="external-link">intervals_to_seq</a></span><span class="op">(</span><span class="va">new_intervals</span><span class="op">)</span></span>
<span><span class="va">seqs</span></span>
<span><span class="co">#&gt; [1] "TTCCTCTCCTCCCCTCGCGCGCGCTCCCTCCTCCCGCAGCCTCTCCTCCACCAGCTGACTCCGAGGGAGAGGATGACCTCATCCCTTCCCTTCCAGCTGCCGCCGCTCCCACCCCGGCTGGGGAGGGGCGAGAAGGAGGGCCCGGAGGAGGGGCTGGGATTGAGGGGAGCGGCGGGTGGGGGTGCCTGGCTGGCCAGTGCTGGACGCGGAGGGCAACAGCACGGCAATCGGAGGCCCAGTCCAGGCTCGTGGGATAGCGAAGAGCGTTGAGTGGATTTCCTCGAAGCTGGGGGGATGGGA"</span></span>
<span><span class="co">#&gt; [2] "ATCTCTGGAAAGACTTGTGCTGATCTCTCTCTGCCCCTTCCTTGATTCACATCTCAAGGGACCGAGAAGGGAGGGAAAACACCAGTCCAGTATTTCCTATCAGTTCAGCGGGGCAGGAACCGGGAGCTTTTCCACAGGGCTGAGCCTGGCCCTCCACTGAGCAGTGTCTGCATTCCAAGGCTCCAGCCTGTCACCACCCTTCCAATCCCTTTGAAGCTGGGCAAAAGGCCTGCCAACAAGCACCAAACTTGAGAGCTCCTCTGCCAGCCCTGGGAGGGGCTGTTTCCTGCCTGCTTTTCG"</span></span>
<span><span class="co">#&gt; [3] "GCTCATGGCTCTCCAGACCGACCCCGAGCTCTGCTATGGCCACGGGACACGCCGCTTCCCCCGACCCTGAGGCAGGGATCGGAAGCTAGCCTGGAGATGCCCAGAGGAACTCGTAAAGCTGAGCGGGTGCTACCCTCCCGCTGCTCTCCTGGTAGACCTAACCCTTCGCCTAATCCGCGCTGGAGATCTACCCAGTGACACTGCGGGTGTCCCCCCGGGCCGCGGGGCCCTTTTCTTTATGGACGCGGCCAATGGCGAGGCGGGGGCGGACCGGACCCTAGTCCTTAGGCCCCCGCCCAG"</span></span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pbm_list.compute.html">pbm_list.compute</a></span><span class="op">(</span><span class="va">pbm_list</span>, <span class="va">seqs</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Computing energies for <span style="color: #0000BB;">21</span> PBMs on <span style="color: #0000BB;">3</span> sequences</span></span>
<span><span class="co">#&gt;          Oct4  Atoh1.2     Tbx5 E2A_1.2     Isl1 E2A_1      MyoG Atoh1.3</span></span>
<span><span class="co">#&gt; [1,] 0.000000 0.000000 3.360820       0 3.729907     0 2.0862625       0</span></span>
<span><span class="co">#&gt; [2,] 2.438457 2.731101 3.491086       0 4.280369     0 2.3392323       0</span></span>
<span><span class="co">#&gt; [3,] 0.000000 0.000000 2.669436       0 3.894363     0 0.5483327       0</span></span>
<span><span class="co">#&gt;      GATA3_2.2    Meis1     E2F6     Pit1    PQM_1  TboxSmad  GATA3_2   E_box_2</span></span>
<span><span class="co">#&gt; [1,]  3.806423 3.154064 4.018626 2.305464 1.271002 0.1399182 6.282286 1.2634071</span></span>
<span><span class="co">#&gt; [2,]  5.520942 6.555252 5.438935 4.640641 5.010974 4.3378987 4.587564 0.4109196</span></span>
<span><span class="co">#&gt; [3,]  4.763735 4.778387 3.007386 4.004477 1.685282 0.4491135 5.887933 0.9321862</span></span>
<span><span class="co">#&gt;          Tcf3      Tcf4    Atoh1 TATA_box Lhx2</span></span>
<span><span class="co">#&gt; [1,] 0.000000 0.4460041 4.839001 0.000000    0</span></span>
<span><span class="co">#&gt; [2,] 2.038721 6.6088638 0.000000 1.872493    0</span></span>
<span><span class="co">#&gt; [3,] 2.091852 0.0000000 0.000000 2.445689    0</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="performance-considerations">Performance Considerations<a class="anchor" aria-label="anchor" href="#performance-considerations"></a>
</h2>
<p>The computational requirements for IceQream depend on the size of your dataset. For reference, the example dataset used here (with ~100,000 peaks and ~20,000 motifs) requires about 23GB of RAM for the motif energy matrix, and will take several hours to create, depending on your hardware.</p>
<p>For a less memory and computationally intensive analysis, you can reduce the number of motifs used in the regression by taking a representative from the SCENIC clusters (1615) instead of all motifs (20,000+). This can be done by:</p>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># motif_energies &lt;- compute_motif_energies(peak_intervals, motif_db_scenic_clusters, normalization_intervals = normalization_intervals)</span></span></code></pre></div>
<p>Note that the performance of the model will be lower with fewer motifs, but it can still provide valuable insights.</p>
</div>
<div class="section level2">
<h2 id="running-iceqream-on-your-own-data">Running IceQream on your own data<a class="anchor" aria-label="anchor" href="#running-iceqream-on-your-own-data"></a>
</h2>
<p>To run IceQream on your own data, you will need to provide the following inputs:</p>
<ol style="list-style-type: decimal">
<li>Genomic positions of peaks (as a dataframe with columns <code>chrom</code>, <code>start</code>, <code>end</code>, <code>peak_name</code>), optionally it can have a <code>const</code> column indicating constitutive loci.</li>
<li>ATAC scores (as a matrix with rows corresponding to peaks and columns corresponding to bins).</li>
<li>(Optional) Additional features (as a data frame with rows corresponding to peaks and columns corresponding to features).</li>
</ol>
<p>You can then follow the steps outlined in this vignette to compute motif energies and run the regression.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aviezer Lifshitz, Akhiad Bercovich, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
