<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Learn 'prego' models for ATAC difference of a trajectory — learn_traj_prego • iceqream</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Learn 'prego' models for ATAC difference of a trajectory — learn_traj_prego"><meta name="description" content="Learn 'prego' models for ATAC difference of a trajectory"><meta property="og:description" content="Learn 'prego' models for ATAC difference of a trajectory"><meta property="og:image" content="https://tanaylab.github.io/iceqream/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iceqream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/iceqream.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Learn 'prego' models for ATAC difference of a trajectory</h1>

      <div class="d-none name"><code>learn_traj_prego.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Learn 'prego' models for ATAC difference of a trajectory</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">learn_traj_prego</span><span class="op">(</span></span>
<span>  <span class="va">peak_intervals</span>,</span>
<span>  <span class="va">atac_diff</span>,</span>
<span>  <span class="va">n_motifs</span>,</span>
<span>  min_diff <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  energy_norm_quantile <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  norm_energy_max <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  min_energy <span class="op">=</span> <span class="op">-</span><span class="fl">7</span>,</span>
<span>  sample_for_kmers <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sample_fraction <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  sequences <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  peaks_size <span class="op">=</span> <span class="fl">300</span>,</span>
<span>  additional_features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  norm_intervals <span class="op">=</span> <span class="va">peak_intervals</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-peak-intervals">peak_intervals<a class="anchor" aria-label="anchor" href="#arg-peak-intervals"></a></dt>
<dd><p>A data frame, indicating the genomic positions ('chrom', 'start', 'end') of each peak, with an additional column named "const" indicating whether the peak is constitutive and therefore shouldn't be used in the regression. Optionally, a column named "cluster" can be added with indication of the cluster of each peak.</p></dd>


<dt id="arg-atac-diff">atac_diff<a class="anchor" aria-label="anchor" href="#arg-atac-diff"></a></dt>
<dd><p>A numeric vector, indicating the ATAC difference of each peak</p></dd>


<dt id="arg-n-motifs">n_motifs<a class="anchor" aria-label="anchor" href="#arg-n-motifs"></a></dt>
<dd><p>Number of motifs to learn. Should be at least 2</p></dd>


<dt id="arg-min-diff">min_diff<a class="anchor" aria-label="anchor" href="#arg-min-diff"></a></dt>
<dd><p>Minimum ATAC difference to include a peak in the training</p></dd>


<dt id="arg-energy-norm-quantile">energy_norm_quantile<a class="anchor" aria-label="anchor" href="#arg-energy-norm-quantile"></a></dt>
<dd><p>quantile of the energy used for normalization. Default: 1</p></dd>


<dt id="arg-norm-energy-max">norm_energy_max<a class="anchor" aria-label="anchor" href="#arg-norm-energy-max"></a></dt>
<dd><p>maximum value of the normalized energy. Default: 10</p></dd>


<dt id="arg-min-energy">min_energy<a class="anchor" aria-label="anchor" href="#arg-min-energy"></a></dt>
<dd><p>Minimum energy value after normalization (default: -7)</p></dd>


<dt id="arg-sample-for-kmers">sample_for_kmers<a class="anchor" aria-label="anchor" href="#arg-sample-for-kmers"></a></dt>
<dd><p>Whether to sample kmers for training. Default: TRUE</p></dd>


<dt id="arg-sample-fraction">sample_fraction<a class="anchor" aria-label="anchor" href="#arg-sample-fraction"></a></dt>
<dd><p>Fraction of peaks to sample for training. Default: 0.1#'</p></dd>


<dt id="arg-sequences">sequences<a class="anchor" aria-label="anchor" href="#arg-sequences"></a></dt>
<dd><p>A character vector of sequences to learn the motifs on. If NULL, the sequences of the peaks are used.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Random seed</p></dd>


<dt id="arg-peaks-size">peaks_size<a class="anchor" aria-label="anchor" href="#arg-peaks-size"></a></dt>
<dd><p>size of the peaks to extract sequences from. Default: 300bp</p></dd>


<dt id="arg-additional-features">additional_features<a class="anchor" aria-label="anchor" href="#arg-additional-features"></a></dt>
<dd><p>A matrix of additional features to filter out before learning the motifs (e.g. CpG content, dinucleotide content, etc.)</p></dd>


<dt id="arg-norm-intervals">norm_intervals<a class="anchor" aria-label="anchor" href="#arg-norm-intervals"></a></dt>
<dd><p>A data frame, indicating the genomic positions ('chrom', 'start', 'end') of peaks used for energy normalization. If NULL, the function will use <code>peak_intervals</code> for normalization.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments to be passed to <code><a href="https://rdrr.io/pkg/prego/man/regress_pwm.html" class="external-link">prego::regress_pwm</a></code></p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aviezer Lifshitz, Akhiad Bercovich, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

