<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Plot IQ Locus — plot_iq_locus • iceqream</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Plot IQ Locus — plot_iq_locus"><meta name="description" content="This function generates a plot of an IQ locus, including DNA sequence, motif responses, and ATAC-seq tracks."><meta property="og:description" content="This function generates a plot of an IQ locus, including DNA sequence, motif responses, and ATAC-seq tracks."><meta property="og:image" content="https://tanaylab.github.io/iceqream/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iceqream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/iceqream.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Plot IQ Locus</h1>

      <div class="d-none name"><code>plot_iq_locus.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function generates a plot of an IQ locus, including DNA sequence, motif responses, and ATAC-seq tracks.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_iq_locus</span><span class="op">(</span></span>
<span>  <span class="va">interval</span>,</span>
<span>  <span class="va">pbm_list</span>,</span>
<span>  <span class="va">atac_tracks</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  ext_width <span class="op">=</span> <span class="fl">200000</span>,</span>
<span>  T_emax <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  T_rmax <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  motifs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_others <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  trim_pbms <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  trim_bits_thresh <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  bits_threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  order_motifs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  atac_names <span class="op">=</span> <span class="va">atac_tracks</span>,</span>
<span>  atac_colors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  atac_sizes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  line_thresh <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  score <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  predicted_score <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dna <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  norm_q <span class="op">=</span> <span class="fl">0.995</span>,</span>
<span>  tracks_q <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  iterator <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  norm_intervals <span class="op">=</span> <span class="fu">gintervals.all</span><span class="op">(</span><span class="op">)</span>,</span>
<span>  atac_smooth <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  ext_atac_smooth <span class="op">=</span> <span class="va">atac_smooth</span> <span class="op">*</span> <span class="fl">5</span>,</span>
<span>  tn5bias_track <span class="op">=</span> <span class="st">"tn5bias"</span>,</span>
<span>  normalize_tn5bias <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tss_intervals <span class="op">=</span> <span class="st">"intervs.global.tss"</span>,</span>
<span>  exon_intervals <span class="op">=</span> <span class="st">"intervs.global.exon"</span>,</span>
<span>  annot_track <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annot_track_name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annot_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#FF9800"</span>, <span class="st">"#009688"</span><span class="op">)</span>,</span>
<span>  annot_track_iterator <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  annot_track_smooth <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  annot_intervals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annot_intervals_color <span class="op">=</span> <span class="st">"darkgreen"</span>,</span>
<span>  mark_conservation <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  conservation_threshold <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  scale_cex <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  dna_height <span class="op">=</span> <span class="fl">0.03</span>,</span>
<span>  filename <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dev <span class="op">=</span> <span class="fu">grDevices</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/grDevices/pdf.html" class="external-link">pdf</a></span>,</span>
<span>  plot_width <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  plot_height <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  base_size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  base_family <span class="op">=</span> <span class="st">"ArialMT"</span>,</span>
<span>  rasterize <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  raster_device <span class="op">=</span> <span class="st">"ragg"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-interval">interval<a class="anchor" aria-label="anchor" href="#arg-interval"></a></dt>
<dd><p>A genomic interval to plot.</p></dd>


<dt id="arg-pbm-list">pbm_list<a class="anchor" aria-label="anchor" href="#arg-pbm-list"></a></dt>
<dd><p>A list of PBM objects.</p></dd>


<dt id="arg-atac-tracks">atac_tracks<a class="anchor" aria-label="anchor" href="#arg-atac-tracks"></a></dt>
<dd><p>A vector of ATAC-seq track names to plot.</p></dd>


<dt id="arg-width">width<a class="anchor" aria-label="anchor" href="#arg-width"></a></dt>
<dd><p>Width of the plot in bp.</p></dd>


<dt id="arg-ext-width">ext_width<a class="anchor" aria-label="anchor" href="#arg-ext-width"></a></dt>
<dd><p>Width of the top plot in bp.</p></dd>


<dt id="arg-t-emax">T_emax<a class="anchor" aria-label="anchor" href="#arg-t-emax"></a></dt>
<dd><p>Numeric, threshold for maximum energy.</p></dd>


<dt id="arg-t-rmax">T_rmax<a class="anchor" aria-label="anchor" href="#arg-t-rmax"></a></dt>
<dd><p>Numeric, threshold for maximum response.</p></dd>


<dt id="arg-motifs">motifs<a class="anchor" aria-label="anchor" href="#arg-motifs"></a></dt>
<dd><p>A list of specific motifs to plot.</p></dd>


<dt id="arg-plot-others">plot_others<a class="anchor" aria-label="anchor" href="#arg-plot-others"></a></dt>
<dd><p>Logical, whether to plot other motifs that pass the threshold when specific motifs are provided.</p></dd>


<dt id="arg-trim-pbms">trim_pbms<a class="anchor" aria-label="anchor" href="#arg-trim-pbms"></a></dt>
<dd><p>Logical, whether to trim PBM objects by removing positions with low information content from the beginning and end of the PSSM.</p></dd>


<dt id="arg-trim-bits-thresh">trim_bits_thresh<a class="anchor" aria-label="anchor" href="#arg-trim-bits-thresh"></a></dt>
<dd><p>Numeric, threshold for trimming PBM objects.</p></dd>


<dt id="arg-bits-threshold">bits_threshold<a class="anchor" aria-label="anchor" href="#arg-bits-threshold"></a></dt>
<dd><p>Numeric, threshold for trimming PSSMs.</p></dd>


<dt id="arg-order-motifs">order_motifs<a class="anchor" aria-label="anchor" href="#arg-order-motifs"></a></dt>
<dd><p>Logical, whether to order motifs by maximum response.</p></dd>


<dt id="arg-atac-names">atac_names<a class="anchor" aria-label="anchor" href="#arg-atac-names"></a></dt>
<dd><p>Character vector, names for ATAC-seq tracks.</p></dd>


<dt id="arg-atac-colors">atac_colors<a class="anchor" aria-label="anchor" href="#arg-atac-colors"></a></dt>
<dd><p>Named vector, colors for ATAC-seq tracks.</p></dd>


<dt id="arg-atac-sizes">atac_sizes<a class="anchor" aria-label="anchor" href="#arg-atac-sizes"></a></dt>
<dd><p>Named vector, line sizes for ATAC-seq tracks.</p></dd>


<dt id="arg-line-thresh">line_thresh<a class="anchor" aria-label="anchor" href="#arg-line-thresh"></a></dt>
<dd><p>Numeric, threshold for drawing vertical lines.</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>Character, optional title for the plot.</p></dd>


<dt id="arg-score">score<a class="anchor" aria-label="anchor" href="#arg-score"></a></dt>
<dd><p>Numeric, optional score to display in the plot title.</p></dd>


<dt id="arg-predicted-score">predicted_score<a class="anchor" aria-label="anchor" href="#arg-predicted-score"></a></dt>
<dd><p>Numeric, optional predicted score to display in the plot title.</p></dd>


<dt id="arg-dna">dna<a class="anchor" aria-label="anchor" href="#arg-dna"></a></dt>
<dd><p>Character, DNA sequence to plot. If NULL, the sequence is extracted from the interval.</p></dd>


<dt id="arg-norm-q">norm_q<a class="anchor" aria-label="anchor" href="#arg-norm-q"></a></dt>
<dd><p>Numeric, quantile for normalization. If a vector, should have the same length as atac_names.</p></dd>


<dt id="arg-tracks-q">tracks_q<a class="anchor" aria-label="anchor" href="#arg-tracks-q"></a></dt>
<dd><p>Data frame, pre-computed quantiles for tracks. Should have a 'type' column and a 'q' column.</p></dd>


<dt id="arg-iterator">iterator<a class="anchor" aria-label="anchor" href="#arg-iterator"></a></dt>
<dd><p>Numeric, iterator for gextract.</p></dd>


<dt id="arg-norm-intervals">norm_intervals<a class="anchor" aria-label="anchor" href="#arg-norm-intervals"></a></dt>
<dd><p>Genomic intervals for normalization.</p></dd>


<dt id="arg-atac-smooth">atac_smooth<a class="anchor" aria-label="anchor" href="#arg-atac-smooth"></a></dt>
<dd><p>Numeric, smoothing factor for ATAC-seq tracks.</p></dd>


<dt id="arg-ext-atac-smooth">ext_atac_smooth<a class="anchor" aria-label="anchor" href="#arg-ext-atac-smooth"></a></dt>
<dd><p>Numeric, smoothing factor for the top plot.</p></dd>


<dt id="arg-tn-bias-track">tn5bias_track<a class="anchor" aria-label="anchor" href="#arg-tn-bias-track"></a></dt>
<dd><p>Character, name of the Tn5 bias track.</p></dd>


<dt id="arg-normalize-tn-bias">normalize_tn5bias<a class="anchor" aria-label="anchor" href="#arg-normalize-tn-bias"></a></dt>
<dd><p>Logical, whether to normalize using the TN5 bias track. If a vector, should have the same length as atac_names.</p></dd>


<dt id="arg-tss-intervals">tss_intervals<a class="anchor" aria-label="anchor" href="#arg-tss-intervals"></a></dt>
<dd><p>Character, name of the TSS intervals track.</p></dd>


<dt id="arg-exon-intervals">exon_intervals<a class="anchor" aria-label="anchor" href="#arg-exon-intervals"></a></dt>
<dd><p>Character, name of the exon intervals track.</p></dd>


<dt id="arg-annot-track">annot_track<a class="anchor" aria-label="anchor" href="#arg-annot-track"></a></dt>
<dd><p>Character, name of the annotation track.</p></dd>


<dt id="arg-annot-track-name">annot_track_name<a class="anchor" aria-label="anchor" href="#arg-annot-track-name"></a></dt>
<dd><p>Character, display name for the annotation track.</p></dd>


<dt id="arg-annot-colors">annot_colors<a class="anchor" aria-label="anchor" href="#arg-annot-colors"></a></dt>
<dd><p>Named vector, colors for annotation track.</p></dd>


<dt id="arg-annot-track-iterator">annot_track_iterator<a class="anchor" aria-label="anchor" href="#arg-annot-track-iterator"></a></dt>
<dd><p>Numeric, iterator for annotation track extraction.</p></dd>


<dt id="arg-annot-track-smooth">annot_track_smooth<a class="anchor" aria-label="anchor" href="#arg-annot-track-smooth"></a></dt>
<dd><p>Numeric, smoothing factor for annotation track.</p></dd>


<dt id="arg-annot-intervals">annot_intervals<a class="anchor" aria-label="anchor" href="#arg-annot-intervals"></a></dt>
<dd><p>Genomic intervals for annotation (start positions would be used).</p></dd>


<dt id="arg-annot-intervals-color">annot_intervals_color<a class="anchor" aria-label="anchor" href="#arg-annot-intervals-color"></a></dt>
<dd><p>Color for annotation intervals.</p></dd>


<dt id="arg-mark-conservation">mark_conservation<a class="anchor" aria-label="anchor" href="#arg-mark-conservation"></a></dt>
<dd><p>Logical, whether to mark conservation in the DNA sequence.</p></dd>


<dt id="arg-conservation-threshold">conservation_threshold<a class="anchor" aria-label="anchor" href="#arg-conservation-threshold"></a></dt>
<dd><p>Numeric, threshold for marking conservation.</p></dd>


<dt id="arg-scale-cex">scale_cex<a class="anchor" aria-label="anchor" href="#arg-scale-cex"></a></dt>
<dd><p>Numeric, scaling factor for letter sizes.</p></dd>


<dt id="arg-dna-height">dna_height<a class="anchor" aria-label="anchor" href="#arg-dna-height"></a></dt>
<dd><p>Numeric, height of each row of DNA sequence relative to the entire plot.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>Character, output filename (if NULL, plot is not saved).</p></dd>


<dt id="arg-dev">dev<a class="anchor" aria-label="anchor" href="#arg-dev"></a></dt>
<dd><p>Function, device to use for plotting.</p></dd>


<dt id="arg-plot-width">plot_width<a class="anchor" aria-label="anchor" href="#arg-plot-width"></a></dt>
<dd><p>Numeric, width of the output plot.</p></dd>


<dt id="arg-plot-height">plot_height<a class="anchor" aria-label="anchor" href="#arg-plot-height"></a></dt>
<dd><p>Numeric, height of the output plot.</p></dd>


<dt id="arg-base-size">base_size<a class="anchor" aria-label="anchor" href="#arg-base-size"></a></dt>
<dd><p>Numeric, base font size for the plot.</p></dd>


<dt id="arg-base-family">base_family<a class="anchor" aria-label="anchor" href="#arg-base-family"></a></dt>
<dd><p>Character, base font family for the plot.</p></dd>


<dt id="arg-rasterize">rasterize<a class="anchor" aria-label="anchor" href="#arg-rasterize"></a></dt>
<dd><p>Logical, whether to rasterize certain plot elements for better performance.</p></dd>


<dt id="arg-raster-device">raster_device<a class="anchor" aria-label="anchor" href="#arg-raster-device"></a></dt>
<dd><p>Character, device to use for rasterization.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>additional arguments to pass to the plotting function.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot object containing the IQ locus plot.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aviezer Lifshitz, Akhiad Bercovich, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

