<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Regress trajectory motifs of a full manifold (multiple trajectories) — regress_trajectory_motifs_manifold • iceqream</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Regress trajectory motifs of a full manifold (multiple trajectories) — regress_trajectory_motifs_manifold"><meta name="description" content="Regress trajectory motifs of a full manifold (multiple trajectories)"><meta property="og:description" content="Regress trajectory motifs of a full manifold (multiple trajectories)"><meta property="og:image" content="https://tanaylab.github.io/iceqream/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">iceqream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/iceqream.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Regress trajectory motifs of a full manifold (multiple trajectories)</h1>

      <div class="d-none name"><code>regress_trajectory_motifs_manifold.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Regress trajectory motifs of a full manifold (multiple trajectories)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">regress_trajectory_motifs_manifold</span><span class="op">(</span></span>
<span>  <span class="va">peak_intervals</span>,</span>
<span>  <span class="va">atac_diff_add_mat</span>,</span>
<span>  norm_intervals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_motif_num <span class="op">=</span> <span class="fl">120</span>,</span>
<span>  target_traj_motif_num <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  n_clust_factor <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  motif_energies <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  norm_motif_energies <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pssm_db <span class="op">=</span> <span class="fu">iceqream</span><span class="fu">::</span><span class="va"><a href="motif_db.html">motif_db</a></span>,</span>
<span>  additional_features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_tss_distance <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  bin_start <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  bin_end <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_initial_energy_cor <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  normalize_energies <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  energy_norm_quantile <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  norm_energy_max <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  n_prego_motifs <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  traj_prego <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  initial_min_diff <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  min_diff <span class="op">=</span> <span class="va">initial_min_diff</span>,</span>
<span>  prego_sample_for_kmers <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  prego_sample_fraction <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">60427</span>,</span>
<span>  feature_selection_beta <span class="op">=</span> <span class="fl">0.003</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">0.00001</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  filter_using_r2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  r2_threshold <span class="op">=</span> <span class="fl">0.0005</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  peaks_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  spat_num_bins <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spat_bin_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  kmer_sequence_length <span class="op">=</span> <span class="fl">300</span>,</span>
<span>  include_interactions <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  interaction_threshold <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  max_motif_interaction_n <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_add_interaction_n <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-peak-intervals">peak_intervals<a class="anchor" aria-label="anchor" href="#arg-peak-intervals"></a></dt>
<dd><p>A data frame, indicating the genomic positions ('chrom', 'start', 'end') of each peak.</p></dd>


<dt id="arg-norm-intervals">norm_intervals<a class="anchor" aria-label="anchor" href="#arg-norm-intervals"></a></dt>
<dd><p>A data frame, indicating the genomic positions ('chrom', 'start', 'end') of peaks used for energy normalization. If NULL, the function will use <code>peak_intervals</code> for normalization.</p></dd>


<dt id="arg-max-motif-num">max_motif_num<a class="anchor" aria-label="anchor" href="#arg-max-motif-num"></a></dt>
<dd><p>maximum number of motifs to consider. Default: 50</p></dd>


<dt id="arg-target-traj-motif-num">target_traj_motif_num<a class="anchor" aria-label="anchor" href="#arg-target-traj-motif-num"></a></dt>
<dd><p>Number of motifs to select for each trajectory. Note that the actual number of motifs can be less or more than this number, depending on the data from other trajectories.</p></dd>


<dt id="arg-n-clust-factor">n_clust_factor<a class="anchor" aria-label="anchor" href="#arg-n-clust-factor"></a></dt>
<dd><p>factor to divide the number of to keep after clustering. e.g. if n_clust_factor &gt; 1 the number of motifs to keep will be reduced by a factor of n_clust_factor. Default: 1</p></dd>


<dt id="arg-motif-energies">motif_energies<a class="anchor" aria-label="anchor" href="#arg-motif-energies"></a></dt>
<dd><p>A numeric matrix, representing the energy of each motif in each peak. If NULL, the function will use <code>pssm_db</code> to calculate the motif energies. Note that this might take a while.</p></dd>


<dt id="arg-norm-motif-energies">norm_motif_energies<a class="anchor" aria-label="anchor" href="#arg-norm-motif-energies"></a></dt>
<dd><p>A numeric matrix, representing the normalized energy of each motif in each interval of <code>norm_intervals</code>. If NULL, the function will use <code>pssm_db</code> to calculate the motif energies. Note that this might take a while.</p></dd>


<dt id="arg-pssm-db">pssm_db<a class="anchor" aria-label="anchor" href="#arg-pssm-db"></a></dt>
<dd><p>a data frame with PSSMs ('A', 'C', 'G' and 'T' columns), with an additional column 'motif' containing the motif name. All the motifs in <code>motif_energies</code> (column names) should be present in the 'motif' column. Default: all motifs.</p></dd>


<dt id="arg-additional-features">additional_features<a class="anchor" aria-label="anchor" href="#arg-additional-features"></a></dt>
<dd><p>A data frame, representing additional genomic features (e.g. CpG content, distance to TSS, etc.) for each peak. Note that NA values would be replaced with 0.</p></dd>


<dt id="arg-min-tss-distance">min_tss_distance<a class="anchor" aria-label="anchor" href="#arg-min-tss-distance"></a></dt>
<dd><p>distance from Transcription Start Site (TSS) to classify a peak as an enhancer. Default: 5000. If NULL, no filtering will be performed - use this option if your peaks are already filtered. <br>
Note that in order to filter peaks that are too close to TSS, the current <code>misha</code> genome must have an intervals set called <code>intervs.global.tss</code>.</p></dd>


<dt id="arg-bin-start">bin_start<a class="anchor" aria-label="anchor" href="#arg-bin-start"></a></dt>
<dd><p>the start of the trajectory. Default: 1</p></dd>


<dt id="arg-bin-end">bin_end<a class="anchor" aria-label="anchor" href="#arg-bin-end"></a></dt>
<dd><p>the end of the trajectory. Default: the last bin (only used when atac_scores is provided)</p></dd>


<dt id="arg-min-initial-energy-cor">min_initial_energy_cor<a class="anchor" aria-label="anchor" href="#arg-min-initial-energy-cor"></a></dt>
<dd><p>minimal correlation between the motif normalized energy and the ATAC difference.</p></dd>


<dt id="arg-normalize-energies">normalize_energies<a class="anchor" aria-label="anchor" href="#arg-normalize-energies"></a></dt>
<dd><p>whether to normalize the motif energies. Set this to FALSE if the motif energies are already normalized.</p></dd>


<dt id="arg-energy-norm-quantile">energy_norm_quantile<a class="anchor" aria-label="anchor" href="#arg-energy-norm-quantile"></a></dt>
<dd><p>quantile of the energy used for normalization. Default: 1</p></dd>


<dt id="arg-norm-energy-max">norm_energy_max<a class="anchor" aria-label="anchor" href="#arg-norm-energy-max"></a></dt>
<dd><p>maximum value of the normalized energy. Default: 10</p></dd>


<dt id="arg-n-prego-motifs">n_prego_motifs<a class="anchor" aria-label="anchor" href="#arg-n-prego-motifs"></a></dt>
<dd><p>number of prego motifs (de-novo motifs) to consider.</p></dd>


<dt id="arg-traj-prego">traj_prego<a class="anchor" aria-label="anchor" href="#arg-traj-prego"></a></dt>
<dd><p>output of <code>learn_traj_prego</code>. If provided, no additional prego models would be inferred.</p></dd>


<dt id="arg-initial-min-diff">initial_min_diff<a class="anchor" aria-label="anchor" href="#arg-initial-min-diff"></a></dt>
<dd><p>minimal ATAC difference for a peak to participate in the initial prego motif inference.</p></dd>


<dt id="arg-min-diff">min_diff<a class="anchor" aria-label="anchor" href="#arg-min-diff"></a></dt>
<dd><p>minimal ATAC difference for a peak to participate in the initial prego motif inference and in the distillation step (if <code>distill_on_diff</code> is TRUE).</p></dd>


<dt id="arg-prego-sample-for-kmers">prego_sample_for_kmers<a class="anchor" aria-label="anchor" href="#arg-prego-sample-for-kmers"></a></dt>
<dd><p>whether to use a sample of the peaks for kmer screening. Default: TRUE</p></dd>


<dt id="arg-prego-sample-fraction">prego_sample_fraction<a class="anchor" aria-label="anchor" href="#arg-prego-sample-fraction"></a></dt>
<dd><p>Fraction of peaks to sample for prego motif inference. A smaller number would be faster but might lead to over-fitting. Default: 0.1</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>random seed for reproducibility.</p></dd>


<dt id="arg-feature-selection-beta">feature_selection_beta<a class="anchor" aria-label="anchor" href="#arg-feature-selection-beta"></a></dt>
<dd><p>beta parameter used for feature selection.</p></dd>


<dt id="arg-lambda">lambda<a class="anchor" aria-label="anchor" href="#arg-lambda"></a></dt>
<dd><p>A user supplied <code>lambda</code> sequence. Typical usage is to
have the program compute its own <code>lambda</code> sequence based on
<code>nlambda</code> and <code>lambda.min.ratio</code>. Supplying a value of
<code>lambda</code> overrides this. WARNING: use with care. Avoid supplying a
single value for <code>lambda</code> (for predictions after CV use
<code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict()</a></code> instead).  Supply instead a decreasing sequence of
<code>lambda</code> values. <code>glmnet</code> relies on its warms starts for speed,
and its often faster to fit a whole path than compute a single fit.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>The elasticnet mixing parameter, with \(0\le\alpha\le 1\).
The penalty is defined as
$$(1-\alpha)/2||\beta||_2^2+\alpha||\beta||_1.$$ <code>alpha=1</code> is the
lasso penalty, and <code>alpha=0</code> the ridge penalty.</p></dd>


<dt id="arg-filter-using-r-">filter_using_r2<a class="anchor" aria-label="anchor" href="#arg-filter-using-r-"></a></dt>
<dd><p>whether to filter features using R^2.</p></dd>


<dt id="arg-r-threshold">r2_threshold<a class="anchor" aria-label="anchor" href="#arg-r-threshold"></a></dt>
<dd><p>minimal R^2 for a feature to be included in the model.</p></dd>


<dt id="arg-parallel">parallel<a class="anchor" aria-label="anchor" href="#arg-parallel"></a></dt>
<dd><p>whether to use parallel processing on glmnet.</p></dd>


<dt id="arg-peaks-size">peaks_size<a class="anchor" aria-label="anchor" href="#arg-peaks-size"></a></dt>
<dd><p>size of the peaks to extract sequences from. Default: 500bp</p></dd>


<dt id="arg-spat-num-bins">spat_num_bins<a class="anchor" aria-label="anchor" href="#arg-spat-num-bins"></a></dt>
<dd><p>number of spatial bins to use.</p></dd>


<dt id="arg-spat-bin-size">spat_bin_size<a class="anchor" aria-label="anchor" href="#arg-spat-bin-size"></a></dt>
<dd><p>size of each spatial bin.</p></dd>


<dt id="arg-kmer-sequence-length">kmer_sequence_length<a class="anchor" aria-label="anchor" href="#arg-kmer-sequence-length"></a></dt>
<dd><p>length of the kmer sequence to use for kmer screening. By default the full sequence is used.</p></dd>


<dt id="arg-include-interactions">include_interactions<a class="anchor" aria-label="anchor" href="#arg-include-interactions"></a></dt>
<dd><p>whether to include interactions between motifs / additional fetures as model features. IQ will create interactions between significant additional features and all motifs, and between significant motifs. Default: FALSE</p></dd>


<dt id="arg-interaction-threshold">interaction_threshold<a class="anchor" aria-label="anchor" href="#arg-interaction-threshold"></a></dt>
<dd><p>threshold for the selecting features to create interactions. IQ learns a linear model on the features and selects the features with coefficients above this threshold. Default: 0.001</p></dd>


<dt id="arg-max-motif-interaction-n">max_motif_interaction_n<a class="anchor" aria-label="anchor" href="#arg-max-motif-interaction-n"></a></dt>
<dd><p>maximum number of motifs to consider for interactions. If NULL, all motifs above the interaction_threshold will be considered. Default: NULL</p></dd>


<dt id="arg-max-add-interaction-n">max_add_interaction_n<a class="anchor" aria-label="anchor" href="#arg-max-add-interaction-n"></a></dt>
<dd><p>maximum number of additional features to consider for interactions. If NULL, all additional features above the interaction_threshold will be considered. Default: NULL</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Arguments passed on to <code><a href="distill_traj_model_multi.html">distill_traj_model_multi</a></code></p><dl><dt><code>traj_models</code></dt>
<dd><p>A list of trajectory models.</p></dd>

    <dt><code>distill_single</code></dt>
<dd><p>Logical indicating whether to distill clusters with a single motif.</p></dd>

    <dt><code>learn_single_spatial</code></dt>
<dd><p>Logical indicating whether to learn only spatial features for clusters with a single motif, or skip them.</p></dd>

    <dt><code>use_all_motifs</code></dt>
<dd><p>Logical indicating whether to use all motifs in the resulting models. If FALSE, only motifs from clusters which had a motif from the original model are used.</p></dd>

    <dt><code>cluster_report_dir</code></dt>
<dd><p>The directory to store cluster reports. If not NULL, a png would be created for each cluster.</p></dd>

    <dt><code>filter_models</code></dt>
<dd><p>Logical indicating whether to filter the models before distillation. Defaults to TRUE.</p></dd>

    <dt><code>unique_motifs</code></dt>
<dd><p>Logical indicating whether to keep only unique motifs. Defaults to FALSE.</p></dd>

    <dt><code>intra_cor_thresh</code></dt>
<dd><p>The threshold for the average intra-cluster correlation to split clusters. If NULL, no splitting is done.</p></dd>

    <dt><code>bits_threshold</code></dt>
<dd><p>minimal sum of bits for a feature to be included in the model.</p></dd>


</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a TrajectoryModelMulti object</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aviezer Lifshitz, Akhiad Bercovich, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

