<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function performs motif regression on ATAC trajectories. It can work with either ATAC scores on trajectory bins or directly with differential accessibility.
The basic inputs for regression are the genomic positions of the peaks, two vectors of ATAC scores (as an atac_scores matrix) or differential accessibility (atac_diff), and database energies computed for all the genomics positions.
Optionally, additional features such as epigenomic features can be provided."><title>Perform motif regression on ATAC trajectories. — regress_trajectory_motifs • iceqream</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform motif regression on ATAC trajectories. — regress_trajectory_motifs"><meta property="og:description" content="This function performs motif regression on ATAC trajectories. It can work with either ATAC scores on trajectory bins or directly with differential accessibility.
The basic inputs for regression are the genomic positions of the peaks, two vectors of ATAC scores (as an atac_scores matrix) or differential accessibility (atac_diff), and database energies computed for all the genomics positions.
Optionally, additional features such as epigenomic features can be provided."><meta property="og:image" content="https://tanaylab.github.io/iceqream/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">iceqream</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/iceqream.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Perform motif regression on ATAC trajectories.</h1>
      
      <div class="d-none name"><code>regress_trajectory_motifs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs motif regression on ATAC trajectories. It can work with either ATAC scores on trajectory bins or directly with differential accessibility.</p>
<p>The basic inputs for regression are the genomic positions of the peaks, two vectors of ATAC scores (as an <code>atac_scores</code> matrix) or differential accessibility (<code>atac_diff</code>), and database energies computed for all the genomics positions.
Optionally, additional features such as epigenomic features can be provided.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">regress_trajectory_motifs</span><span class="op">(</span></span>
<span>  <span class="va">peak_intervals</span>,</span>
<span>  atac_scores <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  atac_diff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize_bins <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  norm_intervals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_motif_num <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  n_clust_factor <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  motif_energies <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  norm_motif_energies <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pssm_db <span class="op">=</span> <span class="va">motif_db</span>,</span>
<span>  additional_features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_tss_distance <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  bin_start <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  bin_end <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_initial_energy_cor <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  normalize_energies <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  energy_norm_quantile <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  norm_energy_max <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  n_prego_motifs <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  traj_prego <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_diff <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  distill_on_diff <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  prego_sample_fraction <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">60427</span>,</span>
<span>  feature_selection_beta <span class="op">=</span> <span class="fl">0.003</span>,</span>
<span>  lambda <span class="op">=</span> <span class="fl">0.00001</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  filter_using_r2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  r2_threshold <span class="op">=</span> <span class="fl">0.0005</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  peaks_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  spat_num_bins <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spat_bin_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  kmer_sequence_length <span class="op">=</span> <span class="fl">300</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>peak_intervals</dt>
<dd><p>A data frame, indicating the genomic positions ('chrom', 'start', 'end') of each peak, with an additional column named "const" indicating whether the peak is constitutive. Optionally, a column named "cluster" can be added with indication of the cluster of each peak.</p></dd>


<dt>atac_scores</dt>
<dd><p>Optional. A numeric matrix, representing mean ATAC score per bin per peak. Rows: peaks, columns: bins. By default iceqream would regress the last column minus the first column. If you want to regress something else, please either set bin_start or bin_end, or provide <code>atac_diff</code> instead. If <code>normalize_bins</code> is TRUE, the scores will be normalized to 0, 1.</p></dd>


<dt>atac_diff</dt>
<dd><p>Optional. A numeric vector representing the differential accessibility between the start and end of the trajectory. Either this or <code>atac_scores</code> must be provided.</p></dd>


<dt>normalize_bins</dt>
<dd><p>whether to normalize the ATAC scores to 0, 1. Default: TRUE</p></dd>


<dt>norm_intervals</dt>
<dd><p>A data frame, indicating the genomic positions ('chrom', 'start', 'end') of peaks used for energy normalization. If NULL, the function will use <code>peak_intervals</code> for normalization.</p></dd>


<dt>max_motif_num</dt>
<dd><p>maximum number of motifs to consider. Default: 50</p></dd>


<dt>n_clust_factor</dt>
<dd><p>factor to divide the number of to keep after clustering. e.g. if n_clust_factor &gt; 1 the number of motifs to keep will be reduced by a factor of n_clust_factor. Default: 1</p></dd>


<dt>motif_energies</dt>
<dd><p>A numeric matrix, representing the energy of each motif in each peak. If NULL, the function will use <code>pssm_db</code> to calculate the motif energies. Note that this might take a while.</p></dd>


<dt>norm_motif_energies</dt>
<dd><p>A numeric matrix, representing the normalized energy of each motif in each interval of <code>norm_intervals</code>. If NULL, the function will use <code>pssm_db</code> to calculate the motif energies. Note that this might take a while.</p></dd>


<dt>pssm_db</dt>
<dd><p>a data frame with PSSMs ('A', 'C', 'G' and 'T' columns), with an additional column 'motif' containing the motif name. All the motifs in <code>motif_energies</code> (column names) should be present in the 'motif' column. Default: all motifs in the prego package.</p></dd>


<dt>additional_features</dt>
<dd><p>A data frame, representing additional genomic features (e.g. CpG content, distance to TSS, etc.) for each peak. Note that NA values would be replaced with 0.</p></dd>


<dt>min_tss_distance</dt>
<dd><p>distance from Transcription Start Site (TSS) to classify a peak as an enhancer. Default: 5000. If NULL, no filtering will be performed - use this option if your peaks are already filtered. <br>
Note that in order to filter peaks that are too close to TSS, the current <code>misha</code> genome must have an intervals set called <code>intervs.global.tss</code>.</p></dd>


<dt>bin_start</dt>
<dd><p>the start of the trajectory. Default: 1</p></dd>


<dt>bin_end</dt>
<dd><p>the end of the trajectory. Default: the last bin (only used when atac_scores is provided)</p></dd>


<dt>min_initial_energy_cor</dt>
<dd><p>minimal correlation between the motif normalized energy and the ATAC difference.</p></dd>


<dt>normalize_energies</dt>
<dd><p>whether to normalize the motif energies. Set this to FALSE if the motif energies are already normalized.</p></dd>


<dt>energy_norm_quantile</dt>
<dd><p>quantile of the energy used for normalization. Default: 1</p></dd>


<dt>norm_energy_max</dt>
<dd><p>maximum value of the normalized energy. Default: 10</p></dd>


<dt>n_prego_motifs</dt>
<dd><p>number of prego motifs (de-novo motifs) to consider.</p></dd>


<dt>traj_prego</dt>
<dd><p>output of <code>learn_traj_prego</code>. If provided, no additional prego models would be inferred.</p></dd>


<dt>min_diff</dt>
<dd><p>minimal ATAC difference for a peak to participate in the initial prego motif inference and in the distillation step (if <code>distill_on_diff</code> is TRUE).</p></dd>


<dt>distill_on_diff</dt>
<dd><p>whether to distill motifs based on differential accessibility. If FALSE, all peaks will be used for distillation, if TRUE - only peaks with differential accessibility &gt;= min_diff will be used.</p></dd>


<dt>prego_sample_fraction</dt>
<dd><p>Fraction of peaks to sample for prego motif inference. A smaller number would be faster but might lead to over-fitting. Default: 0.1</p></dd>


<dt>seed</dt>
<dd><p>random seed for reproducibility.</p></dd>


<dt>feature_selection_beta</dt>
<dd><p>beta parameter used for feature selection.</p></dd>


<dt>lambda</dt>
<dd><p>A user supplied <code>lambda</code> sequence. Typical usage is to
have the program compute its own <code>lambda</code> sequence based on
<code>nlambda</code> and <code>lambda.min.ratio</code>. Supplying a value of
<code>lambda</code> overrides this. WARNING: use with care. Avoid supplying a
single value for <code>lambda</code> (for predictions after CV use
<code><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict()</a></code> instead).  Supply instead a decreasing sequence of
<code>lambda</code> values. <code>glmnet</code> relies on its warms starts for speed,
and its often faster to fit a whole path than compute a single fit.</p></dd>


<dt>alpha</dt>
<dd><p>The elasticnet mixing parameter, with \(0\le\alpha\le 1\).
The penalty is defined as
$$(1-\alpha)/2||\beta||_2^2+\alpha||\beta||_1.$$ <code>alpha=1</code> is the
lasso penalty, and <code>alpha=0</code> the ridge penalty.</p></dd>


<dt>filter_using_r2</dt>
<dd><p>whether to filter features using R^2.</p></dd>


<dt>r2_threshold</dt>
<dd><p>minimal R^2 for a feature to be included in the model.</p></dd>


<dt>parallel</dt>
<dd><p>whether to use parallel processing on glmnet.</p></dd>


<dt>peaks_size</dt>
<dd><p>size of the peaks to extract sequences from. Default: 300bp</p></dd>


<dt>spat_num_bins</dt>
<dd><p>number of spatial bins to use.</p></dd>


<dt>spat_bin_size</dt>
<dd><p>size of each spatial bin.</p></dd>


<dt>kmer_sequence_length</dt>
<dd><p>length of the kmer sequence to use for kmer screening. By default the full sequence is used.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>An instance of <code>TrajectoryModel</code> containing model information and results:</p><ul><li><p>modelThe final General Linear Model (GLM) object.</p></li>
<li><p>motif_modelsNamed List, PSSM and spatial models for each motif cluster.</p></li>
<li><p>normalized_energiesNumeric vector, normalized energies of each motif in each peak.</p></li>
<li><p>additional_featuresdata frame of the additional features.</p></li>
<li><p>diff_scoreNumeric, normalized score of differential accessibility between 'bin_start' and 'bin_end'.</p></li>
<li><p>predicted_diff_scoreNumeric, predicted differential accessibility score between 'bin_start' and 'bin_end'.</p></li>
<li><p>initial_prego_modelsList, inferred prego models at the initial step of the algorithm.</p></li>
<li><p>peak_intervalsdata frame, indicating the genomic positions ('chrom', 'start', 'end') of each peak used for training.</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aviezer Lifshitz, Akhiad Bercovich, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

